// @klyra/chain/dist/src/utils/chain-helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.round = round;
exports.calculateQuantums = calculateQuantums;
exports.calculateVaultQuantums = calculateVaultQuantums;
exports.calculateSubticks = calculateSubticks;
exports.calculateSide = calculateSide;
exports.calculateTimeInForce = calculateTimeInForce;
exports.calculateOrderFlags = calculateOrderFlags;
exports.calculateClientMetadata = calculateClientMetadata;
exports.calculateConditionType = calculateConditionType;
exports.calculateConditionalOrderTriggerSubticks = calculateConditionalOrderTriggerSubticks;
const long_1 = __importDefault(require("long"));
const shared_1 = require("@klyra/shared");
const proto_includes_1 = require("../lib/proto-includes");
const types_1 = require("../types");
function round(input, base) {
    const rounded = Math.floor(input / base) * base;
    return Number(rounded.toFixed(10)); // Use toFixed to avoid floating point precision issues
}
function calculateQuantums(size, atomicResolution, stepBaseQuantums) {
    const rawQuantums = size * Math.pow(10, -atomicResolution);
    const quantums = round(rawQuantums, stepBaseQuantums);
    // stepBaseQuantums functions as minimum order size
    const result = Math.max(quantums, stepBaseQuantums);
    return long_1.default.fromNumber(result);
}
function calculateVaultQuantums(size) {
    return BigInt((size * 1000000).toFixed(0));
}
function calculateSubticks(price, atomicResolution, quantumConversionExponent, subticksPerTick) {
    const QUOTE_QUANTUMS_ATOMIC_RESOLUTION = -6;
    const exponent = atomicResolution -
        quantumConversionExponent -
        QUOTE_QUANTUMS_ATOMIC_RESOLUTION;
    const rawSubticks = price * Math.pow(10, exponent);
    const subticks = round(rawSubticks, subticksPerTick);
    const result = Math.max(subticks, subticksPerTick);
    return long_1.default.fromNumber(result);
}
function calculateSide(side) {
    return side === shared_1.OrderSide.BUY ? proto_includes_1.Order_Side.SIDE_BUY : proto_includes_1.Order_Side.SIDE_SELL;
}
function calculateTimeInForce(type, timeInForce, execution, postOnly) {
    switch (type) {
        case shared_1.OrderType.MARKET:
            switch (timeInForce) {
                default:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_IOC;
            }
        case shared_1.OrderType.LIMIT:
            switch (timeInForce) {
                case shared_1.OrderTimeInForce.GTT:
                    if (typeof postOnly === "undefined") {
                        throw new Error("postOnly must be set if order type is LIMIT and timeInForce is GTT");
                    }
                    return postOnly
                        ? proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_POST_ONLY
                        : proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_UNSPECIFIED;
                case shared_1.OrderTimeInForce.IOC:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_IOC;
                default:
                    throw new Error("Unexpected code path: timeInForce");
            }
        case shared_1.OrderType.STOP_LIMIT:
        case shared_1.OrderType.TAKE_PROFIT_LIMIT:
            if (typeof execution === "undefined") {
                throw new Error("execution must be set if order type is STOP_LIMIT or TAKE_PROFIT_LIMIT");
            }
            switch (execution) {
                case shared_1.OrderExecution.DEFAULT:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_UNSPECIFIED;
                case shared_1.OrderExecution.POST_ONLY:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_POST_ONLY;
                case shared_1.OrderExecution.IOC:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_IOC;
                default:
                    throw new Error("Unexpected code path: timeInForce");
            }
        case shared_1.OrderType.STOP_MARKET:
        case shared_1.OrderType.TAKE_PROFIT_MARKET:
            if (typeof execution === "undefined") {
                throw new Error("execution must be set if order type is STOP_MARKET or TAKE_PROFIT_MARKET");
            }
            switch (execution) {
                case shared_1.OrderExecution.DEFAULT:
                    throw new Error("Execution value DEFAULT not supported for STOP_MARKET or TAKE_PROFIT_MARKET");
                case shared_1.OrderExecution.POST_ONLY:
                    throw new Error("Execution value POST_ONLY not supported for STOP_MARKET or TAKE_PROFIT_MARKET");
                case shared_1.OrderExecution.IOC:
                    return proto_includes_1.Order_TimeInForce.TIME_IN_FORCE_IOC;
                default:
                    throw new Error("Unexpected code path: timeInForce");
            }
        default:
            throw new Error("Unexpected code path: timeInForce");
    }
}
function calculateOrderFlags(type, timeInForce) {
    switch (type) {
        case shared_1.OrderType.MARKET:
            return types_1.OrderFlags.SHORT_TERM;
        case shared_1.OrderType.LIMIT:
            if (timeInForce === undefined) {
                throw new Error("timeInForce must be set if orderType is LIMIT");
            }
            if (timeInForce === shared_1.OrderTimeInForce.GTT) {
                return types_1.OrderFlags.LONG_TERM;
            }
            return types_1.OrderFlags.SHORT_TERM;
        case shared_1.OrderType.STOP_LIMIT:
            return types_1.OrderFlags.CONDITIONAL;
        case shared_1.OrderType.TAKE_PROFIT_LIMIT:
            return types_1.OrderFlags.CONDITIONAL;
        case shared_1.OrderType.STOP_MARKET:
            return types_1.OrderFlags.CONDITIONAL;
        case shared_1.OrderType.TAKE_PROFIT_MARKET:
            return types_1.OrderFlags.CONDITIONAL;
        default:
            throw new Error("Unexpected code path: orderFlags");
    }
}
function calculateClientMetadata(orderType) {
    switch (orderType) {
        case shared_1.OrderType.MARKET:
            return 1;
        case shared_1.OrderType.STOP_MARKET:
            return 1;
        case shared_1.OrderType.TAKE_PROFIT_MARKET:
            return 1;
        default:
            return 0;
    }
}
function calculateConditionType(orderType) {
    switch (orderType) {
        case shared_1.OrderType.LIMIT:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_UNSPECIFIED;
        case shared_1.OrderType.MARKET:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_UNSPECIFIED;
        case shared_1.OrderType.STOP_LIMIT:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_STOP_LOSS;
        case shared_1.OrderType.STOP_MARKET:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_STOP_LOSS;
        case shared_1.OrderType.TAKE_PROFIT_LIMIT:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_TAKE_PROFIT;
        case shared_1.OrderType.TAKE_PROFIT_MARKET:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_TAKE_PROFIT;
        default:
            return proto_includes_1.Order_ConditionType.CONDITION_TYPE_UNSPECIFIED;
    }
}
function calculateConditionalOrderTriggerSubticks(orderType, atomicResolution, quantumConversionExponent, subticksPerTick, triggerPrice) {
    switch (orderType) {
        case shared_1.OrderType.LIMIT:
            return long_1.default.fromNumber(0);
        case shared_1.OrderType.MARKET:
            return long_1.default.fromNumber(0);
        case shared_1.OrderType.STOP_LIMIT:
        case shared_1.OrderType.STOP_MARKET:
        case shared_1.OrderType.TAKE_PROFIT_LIMIT:
        case shared_1.OrderType.TAKE_PROFIT_MARKET:
            return maybeGetSubticks(atomicResolution, quantumConversionExponent, subticksPerTick, triggerPrice);
        default:
            return long_1.default.fromNumber(0);
    }
}
function maybeGetSubticks(atomicResolution, quantumConversionExponent, subticksPerTick, triggerPrice) {
    if (triggerPrice === undefined) {
        throw new Error("triggerPrice must be set if orderType is STOP_LIMIT, STOP_MARKET, TAKE_PROFIT_LIMIT, or TAKE_PROFIT_MARKET");
    }
    return calculateSubticks(triggerPrice, atomicResolution, quantumConversionExponent, subticksPerTick);
}
