// @klyra/chain/dist/src/utils/helpers.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ByteArrayEncoding = void 0;
exports.convertPartialTransactionOptionsToFull = convertPartialTransactionOptionsToFull;
exports.stripHexPrefix = stripHexPrefix;
exports.bigIntToBytes = bigIntToBytes;
exports.encodeJson = encodeJson;
const encoding_1 = require("@cosmjs/encoding");
const long_1 = __importDefault(require("long"));
const constants_1 = require("../constants");
/**
 * @description Either return undefined or insert default sequence value into
 * `partialTransactionOptions` if it does not exist.
 *
 * @returns undefined or full TransactionOptions.
 */
function convertPartialTransactionOptionsToFull(partialTransactionOptions) {
    if (partialTransactionOptions === undefined) {
        return undefined;
    }
    return {
        sequence: constants_1.DEFAULT_SEQUENCE,
        ...partialTransactionOptions,
    };
}
/**
 * @description Strip '0x' prefix from input string. If there is no '0x' prefix, return the original
 * input.
 *
 * @returns input without '0x' prefix or original input if no prefix.
 */
function stripHexPrefix(input) {
    if (input.startsWith("0x")) {
        return input.slice(2);
    }
    return input;
}
function toBigInt(u) {
    if (u.length <= 1) {
        return BigInt(0);
    }
    // eslint-disable-next-line no-bitwise -- bitwise operations are needed here
    const negated = (u[0] & 1) === 1;
    const hex = Array.from(u.slice(1))
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
    const abs = BigInt(`0x${hex}`);
    return negated ? -abs : abs;
}
function bigIntToBytes(value) {
    const absoluteValue = value < 0 ? value * BigInt(-1) : value;
    const nonPaddedHexValue = absoluteValue.toString(16);
    const paddedHexValue = nonPaddedHexValue.length % 2 === 0
        ? nonPaddedHexValue
        : `0${nonPaddedHexValue}`;
    const numberBytes = new Uint8Array(paddedHexValue.match(/.{1,2}/g).map((byte) => parseInt(byte, 16)));
    const signedBytes = new Uint8Array([value < 0 ? 3 : 2, ...numberBytes]);
    return signedBytes;
}
var ByteArrayEncoding;
(function (ByteArrayEncoding) {
    ByteArrayEncoding["HEX"] = "hex";
    ByteArrayEncoding["BIGINT"] = "bigint";
})(ByteArrayEncoding || (exports.ByteArrayEncoding = ByteArrayEncoding = {}));
/* eslint-disable @typescript-eslint/no-unsafe-member-access -- any and unsafe member access needed for JSON replacer function */
/* eslint-disable @typescript-eslint/no-unsafe-argument -- unsafe argument needed for JSON replacer function */
/* eslint-disable @typescript-eslint/no-unsafe-return -- any is needed for JSON replacer function */
// TODO: Add better type safety for this function
function encodeJson(object, byteArrayEncoding = ByteArrayEncoding.HEX) {
    return JSON.stringify(object, function replacer(_key, value) {
        // Even though we set the an UInt8Array as the value,
        // it comes in here as an object with UInt8Array as the buffer property.
        if (typeof value === "bigint") {
            return value.toString();
        }
        if (value instanceof long_1.default) {
            return value.toString();
        }
        if ((value === null || value === void 0 ? void 0 : value.buffer) instanceof Uint8Array) {
            if (byteArrayEncoding === ByteArrayEncoding.HEX) {
                return (0, encoding_1.toHex)(value.buffer);
            }
            return toBigInt(value.buffer).toString();
        }
        else if (value instanceof Uint8Array) {
            if (byteArrayEncoding === ByteArrayEncoding.HEX) {
                return (0, encoding_1.toHex)(value);
            }
            return toBigInt(value).toString();
        }
        return value;
    });
}
