// @klyra/proto/src/codegen/klyraprotocol/ve/ve.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DaemonVoteExtension = exports.PricePair = void 0;
const _m0 = __importStar(require("protobufjs/minimal"));
function createBasePricePair() {
    return {
        marketId: 0,
        spotPrice: new Uint8Array(),
        pnlPrice: new Uint8Array()
    };
}
exports.PricePair = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.marketId !== 0) {
            writer.uint32(8).uint32(message.marketId);
        }
        if (message.spotPrice.length !== 0) {
            writer.uint32(18).bytes(message.spotPrice);
        }
        if (message.pnlPrice.length !== 0) {
            writer.uint32(26).bytes(message.pnlPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBasePricePair();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketId = reader.uint32();
                    break;
                case 2:
                    message.spotPrice = reader.bytes();
                    break;
                case 3:
                    message.pnlPrice = reader.bytes();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBasePricePair();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : 0;
        message.spotPrice = (_b = object.spotPrice) !== null && _b !== void 0 ? _b : new Uint8Array();
        message.pnlPrice = (_c = object.pnlPrice) !== null && _c !== void 0 ? _c : new Uint8Array();
        return message;
    }
};
function createBaseDaemonVoteExtension() {
    return {
        prices: [],
        sDaiConversionRate: ""
    };
}
exports.DaemonVoteExtension = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.prices) {
            exports.PricePair.encode(v, writer.uint32(10).fork()).ldelim();
        }
        if (message.sDaiConversionRate !== "") {
            writer.uint32(18).string(message.sDaiConversionRate);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseDaemonVoteExtension();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.prices.push(exports.PricePair.decode(reader, reader.uint32()));
                    break;
                case 2:
                    message.sDaiConversionRate = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseDaemonVoteExtension();
        message.prices = ((_a = object.prices) === null || _a === void 0 ? void 0 : _a.map(e => exports.PricePair.fromPartial(e))) || [];
        message.sDaiConversionRate = (_b = object.sDaiConversionRate) !== null && _b !== void 0 ? _b : "";
        return message;
    }
};
