// @klyra/cortex/dist/src/processors/account/account.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountProcessor = void 0;
const subaccount_processor_1 = require("./subaccount.processor");
const account_balances_processor_1 = require("./account-balances.processor");
class AccountProcessor {
    constructor(store) {
        this.store = store;
        this.subaccountProcessor = new subaccount_processor_1.SubaccountProcessor(store);
        this.accountBalancesProcessor = new account_balances_processor_1.AccountBalancesProcessor(store);
    }
    // WebSocket Methods
    processSubscribed(msg) {
        // TODO: update other fields besides subaccounts (orders, equity, free collateral, etc)
        const payload = msg.contents;
        const orders = payload.orders;
        const subaccounts = payload.subaccount.childSubaccounts;
        const newSubaccounts = {};
        for (const subaccount of subaccounts) {
            const newSubaccount = this.subaccountProcessor.process(subaccount, orders);
            if (typeof newSubaccount !== "undefined" && newSubaccount !== null) {
                newSubaccounts[newSubaccount.subaccountNumber] = newSubaccount;
            }
        }
        this.updateSubaccounts(newSubaccounts);
    }
    processBatchUpdate(msg) {
        var _a;
        const payload = msg.contents;
        const subaccountNumber = Number((_a = msg.id) === null || _a === void 0 ? void 0 : _a.split("/")[1]);
        if (typeof subaccountNumber === "undefined" || isNaN(subaccountNumber)) {
            return;
        }
        const existingSubaccounts = this.store.getState().user.account.subaccounts;
        const existingSubaccount = existingSubaccounts[subaccountNumber];
        const newSubaccount = this.subaccountProcessor.processUpdate(existingSubaccount, payload, subaccountNumber);
        if (newSubaccount) {
            this.updateSubaccount(newSubaccount);
        }
    }
    // REST Methods
    //   public processAccountBalances(msg: IndexerAccountBalancesMessage): void {}
    updateSubaccount(subaccount) {
        const existingSubaccounts = this.store.getState().user.account.subaccounts;
        const newSubaccounts = { ...existingSubaccounts };
        newSubaccounts[subaccount.subaccountNumber] = subaccount;
        this.updateSubaccounts(newSubaccounts);
    }
    updateSubaccounts(subaccounts) {
        const current = this.store.getState().user;
        const newUser = {
            ...current,
            account: {
                ...current.account,
                subaccounts,
            },
        };
        this.store.setState({ user: newUser });
    }
}
exports.AccountProcessor = AccountProcessor;
