// @klyra/cortex/dist/src/processors/account/fills.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FillsProcessor = void 0;
const indexer_conversion_1 = require("../../utils/indexer-conversion");
class FillsProcessor {
    constructor(store) {
        this.store = store;
    }
    process(existing, payload, subaccountNumber) {
        if (!payload)
            return [];
        const newFills = payload.map((fill) => this.processFill(fill, subaccountNumber));
        const result = new Set([...existing, ...newFills]);
        const resultsArray = Array.from(result).filter((fill) => fill !== null);
        return resultsArray;
    }
    processFill(fill, subaccountNumber) {
        if (!fill.id)
            return null;
        return {
            id: fill.id,
            marketId: fill.ticker,
            orderId: fill.orderId,
            subaccountNumber,
            marginMode: (0, indexer_conversion_1.getMarginModeFromSubaccount)(subaccountNumber),
            side: (0, indexer_conversion_1.convertIndexerOrderSide)(fill.side),
            type: (0, indexer_conversion_1.convertIndexerOrderType)(fill.type),
            liquidity: (0, indexer_conversion_1.convertIndexerLiquidity)(fill.liquidity),
            price: Number(fill.price),
            size: Number(fill.size),
            fee: Number(fill.fee),
            createdAtMilliseconds: new Date(fill.createdAt).getTime(),
        };
    }
}
exports.FillsProcessor = FillsProcessor;
