// @klyra/cortex/dist/src/processors/account/orders.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrdersProcessor = void 0;
const shared_1 = require("@klyra/shared");
const market_id_1 = require("../../utils/market-id");
const indexer_conversion_1 = require("../../utils/indexer-conversion");
const order_status_1 = require("../../utils/order-status");
class OrdersProcessor {
    constructor(store) {
        this.store = store;
    }
    process(subaccountOrders, payload, subaccountNumber) {
        if (!payload)
            return [];
        const newOrders = payload
            .map((item) => {
            var _a;
            const id = item.id || item.clientId;
            return this.processOrder((_a = subaccountOrders.find((order) => order.id === id)) !== null && _a !== void 0 ? _a : null, item, subaccountNumber);
        })
            .filter((order) => order !== null);
        return newOrders;
    }
    processOrder(existing, payload, subaccountNumber) {
        var _a, _b, _c;
        if (!this.shouldUpdate(existing, payload)) {
            return existing;
        }
        const id = payload.id;
        if (!id)
            return null;
        const type = (0, indexer_conversion_1.convertIndexerOrderType)(payload.type);
        const side = (0, indexer_conversion_1.convertIndexerOrderSide)(payload.side);
        const status = (0, indexer_conversion_1.convertIndexerOrderStatus)(payload.status);
        const marketId = payload.ticker;
        const price = Number(payload.price);
        const size = Number(payload.size);
        if (!marketId || !price || !size) {
            return null;
        }
        const orderFlags = Number(payload.orderFlags);
        const cancelReason = payload.removalReason;
        const totalFilled = Number(payload.totalFilled) || 0;
        const remainingSize = size - totalFilled;
        const updatedAtMilliseconds = payload.updatedAt
            ? new Date(payload.updatedAt).getTime()
            : undefined;
        let modifiedStatus = status;
        if (totalFilled !== 0 && remainingSize !== 0) {
            if (status.value === shared_1.OrderStatusEnum.OPEN) {
                modifiedStatus = (0, order_status_1.getOrderStatusFromEnum)(shared_1.OrderStatusEnum.PARTIALLY_FILLED);
            }
            else if (status.value === shared_1.OrderStatusEnum.CANCELLED) {
                modifiedStatus = (0, order_status_1.getOrderStatusFromEnum)(shared_1.OrderStatusEnum.PARTIALLY_CANCELLED);
            }
        }
        if (orderFlags) {
            const isShortTermOrder = orderFlags === 0;
            const isBestEffortCancelled = modifiedStatus.value === shared_1.OrderStatusEnum.CANCELLING;
            // TODO: add types to the cancel reason
            const isUserCancelled = cancelReason === "USER_CANCELLED" ||
                cancelReason === "ORDER_REMOVAL_REASON_USER_CANCELLED";
            if (isShortTermOrder && isBestEffortCancelled && !isUserCancelled) {
                modifiedStatus = (0, order_status_1.getOrderStatusFromEnum)(shared_1.OrderStatusEnum.PENDING);
            }
        }
        const marginMode = (0, indexer_conversion_1.getMarginModeFromSubaccount)(subaccountNumber);
        const assetId = (0, market_id_1.getAssetIdFromMarketId)(marketId);
        const order = {
            subaccountNumber,
            id,
            assetId,
            clientId: Number(payload.clientId),
            type,
            side,
            status: modifiedStatus,
            timeInForce: (_a = (0, indexer_conversion_1.convertIndexerTimeInForce)(payload.timeInForce)) !== null && _a !== void 0 ? _a : undefined,
            marketId,
            clobPairId: Number(payload.clobPairId),
            price,
            size,
            triggerPrice: Number(payload.triggerPrice),
            trailingPercent: undefined,
            remainingSize,
            totalFilled,
            goodTilBlock: Number(payload.goodTilBlock),
            goodTilBlockTime: Math.floor(new Date((_b = payload.goodTilBlockTime) !== null && _b !== void 0 ? _b : 0).getTime() / 1000), // in seconds
            createdAtHeight: Number(payload.createdAtHeight),
            createdAtMilliseconds: undefined,
            unfillableAtMilliseconds: undefined,
            expiresAtMilliseconds: new Date((_c = payload.goodTilBlockTime) !== null && _c !== void 0 ? _c : 0).getTime(),
            updatedAtMilliseconds,
            postOnly: Boolean(payload.postOnly),
            reduceOnly: Boolean(payload.reduceOnly),
            cancelReason,
            marginMode,
        };
        return order;
    }
    shouldUpdate(existing, payload) {
        var _a, _b;
        const existingDate = (_a = existing === null || existing === void 0 ? void 0 : existing.updatedAtMilliseconds) !== null && _a !== void 0 ? _a : existing === null || existing === void 0 ? void 0 : existing.createdAtMilliseconds;
        const updatedAt = existingDate ? new Date(existingDate) : undefined;
        const incomingUpdatedAt = payload.updatedAt
            ? new Date(payload.updatedAt)
            : undefined;
        if (updatedAt) {
            if (incomingUpdatedAt) {
                if (updatedAt < incomingUpdatedAt) {
                    return true;
                }
                else if (updatedAt > incomingUpdatedAt) {
                    return false;
                }
            }
        }
        else if (incomingUpdatedAt) {
            return true;
        }
        const filled = (_b = existing === null || existing === void 0 ? void 0 : existing.totalFilled) !== null && _b !== void 0 ? _b : 0;
        const incomingFilled = Number(payload.totalFilled);
        if (incomingFilled > filled) {
            return true;
        }
        else if (incomingFilled < filled) {
            return false;
        }
        // If the existing order is cancelling, we need to check if the incoming order is filled, canceled, or best effort canceled
        const finalStatuses = [
            shared_1.IndexerOrderStatus.FILLED,
            shared_1.IndexerOrderStatus.CANCELED,
            shared_1.IndexerOrderStatus.BEST_EFFORT_CANCELED,
        ];
        if ((existing === null || existing === void 0 ? void 0 : existing.status.value) === shared_1.OrderStatusEnum.CANCELLING) {
            return finalStatuses.includes(payload.status);
        }
        return (existing === null || existing === void 0 ? void 0 : existing.status.isFinalized) === false;
    }
}
exports.OrdersProcessor = OrdersProcessor;
