// @klyra/cortex/dist/src/processors/account/perpetual-positions.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PerpetualPositionsProcessor = void 0;
const shared_1 = require("@klyra/shared");
const accounts_1 = require("../../types/accounts");
const market_id_1 = require("../../utils/market-id");
const indexer_conversion_1 = require("../../utils/indexer-conversion");
class PerpetualPositionsProcessor {
    constructor(store) {
        this.store = store;
    }
    process(existing, payload) {
        var _a;
        if (!payload)
            return existing;
        const result = {};
        for (const [market, value] of Object.entries(payload)) {
            const existingPosition = (_a = existing[market]) !== null && _a !== void 0 ? _a : {};
            const newPosition = this.processPerpetualPosition(existingPosition, value);
            if (newPosition) {
                result[market] = newPosition;
            }
            else {
                // delete the key
                const { [market]: _, ...rest } = result;
                return rest;
            }
        }
        return result !== existing ? result : existing;
    }
    processChanges(existing, payload) {
        var _a;
        if (!payload)
            return existing;
        const result = existing;
        for (const item of payload) {
            if (item.market) {
                const newPosition = this.processPerpetualPosition((_a = existing[item.market]) !== null && _a !== void 0 ? _a : {}, 
                // TODO: find a better way to convert the payload (or don't convert at all and improve the processPerpetualPosition method)
                this.convertPerpetualPosition(item));
                if (newPosition) {
                    result[item.market] = newPosition;
                }
                else {
                    // delete the key
                    const { [item.market]: _, ...rest } = result;
                    return rest;
                }
            }
        }
        return result !== existing ? result : existing;
    }
    processPerpetualPosition(existing, payload) {
        var _a;
        if (payload.status === shared_1.IndexerPerpetualPositionStatus.CLOSED) {
            return null;
        }
        const modified = { ...existing };
        const marketId = payload.market;
        const status = payload.status;
        const entryPrice = Number(payload.entryPrice);
        const size = Number(payload.size);
        const isOpen = status === shared_1.IndexerPerpetualPositionStatus.OPEN;
        if (size && payload.status in shared_1.IndexerPerpetualPositionStatus) {
            const realizedPnl = Number(payload.realizedPnl);
            if (typeof realizedPnl === "number") {
                if (isOpen) {
                    // if the position is not closed, we need to calculate the realized PNL percent
                    if (entryPrice) {
                        const positionEntryValue = Math.abs(entryPrice * size);
                        const realizedPnlPercent = Math.abs(realizedPnl / positionEntryValue);
                        modified.realizedPnlPercent =
                            realizedPnlPercent > 0 ? realizedPnlPercent : null;
                    }
                }
                else {
                    modified.realizedPnlPercent = null;
                }
            }
            else {
                modified.realizedPnlPercent = null;
            }
        }
        const storeMarket = this.store.getState().markets[marketId];
        const marketSpotPrice = storeMarket === null || storeMarket === void 0 ? void 0 : storeMarket.spotPrice;
        if (marketSpotPrice) {
            if (isOpen) {
                const valueTotal = marketSpotPrice * size;
                const notional = Math.abs(valueTotal);
                const adjustedImf = storeMarket.configs.effectiveInitialMarginFraction || 0;
                const adjustedMmf = storeMarket.configs.maintenanceMarginFraction || 0;
                const maxLeverage = adjustedImf !== 0 ? 1 / adjustedImf : null;
                const initialRiskTotal = notional * adjustedImf;
                if (entryPrice) {
                    const entryValue = size * entryPrice;
                    const currentValue = size * marketSpotPrice;
                    const unrealizedPnl = currentValue - entryValue;
                    const unrealizedPnlPercent = entryValue !== 0 ? unrealizedPnl / Math.abs(entryValue) : null;
                    modified.unrealizedPnl = unrealizedPnl;
                    modified.unrealizedPnlPercent = unrealizedPnlPercent;
                }
                const marginMode = (_a = existing.marginMode) !== null && _a !== void 0 ? _a : (0, indexer_conversion_1.getMarginModeFromSubaccount)(payload.subaccountNumber);
                let marginValue = null;
                if (marginMode === accounts_1.MarginMode.CROSS) {
                    const maintenanceMargin = storeMarket.configs.maintenanceMarginFraction || 0;
                    marginValue = maintenanceMargin * notional;
                    modified.marginValue = marginValue;
                }
                else {
                    const subaccount = this.store.getState().user.account.subaccounts[payload.subaccountNumber];
                    const equity = Number(subaccount === null || subaccount === void 0 ? void 0 : subaccount.equity);
                    marginValue = equity;
                }
                modified.valueTotal = valueTotal;
                modified.notionalTotal = notional;
                modified.adjustedImf = adjustedImf;
                modified.adjustedMmf = adjustedMmf;
                modified.initialRiskTotal = initialRiskTotal;
                modified.maxLeverage = maxLeverage;
                modified.marginValue = marginValue;
            }
            else {
                modified.valueTotal = null;
                modified.notionalTotal = null;
                modified.adjustedImf = null;
                modified.adjustedMmf = null;
                modified.initialRiskTotal = null;
                modified.maxLeverage = null;
                modified.unrealizedPnl = null;
                modified.unrealizedPnlPercent = null;
                modified.marginValue = null;
            }
        }
        else {
            modified.realizedPnlPercent = null;
            modified.unrealizedPnl = null;
            modified.unrealizedPnlPercent = null;
            modified.valueTotal = null;
            modified.notionalTotal = null;
            modified.adjustedImf = null;
            modified.adjustedMmf = null;
            modified.initialRiskTotal = null;
            modified.maxLeverage = null;
            modified.marginValue = null;
        }
        const side = (0, indexer_conversion_1.convertIndexerPositionSide)(payload.side);
        const assetId = (0, market_id_1.getAssetIdFromMarketId)(marketId);
        const transformed = {
            ...existing,
            ...modified,
            id: payload.market,
            assetId: assetId !== null && assetId !== void 0 ? assetId : "",
            size: Number(payload.size),
            side,
            entryPrice,
            createdAtMilliseconds: payload.createdAt
                ? new Date(payload.createdAt).getTime()
                : undefined,
            exitPrice: payload.exitPrice ? Number(payload.exitPrice) : undefined,
        };
        return transformed;
    }
    convertPerpetualPosition(payload) {
        return {
            ...payload,
            createdAt: "",
            createdAtHeight: "",
        };
    }
}
exports.PerpetualPositionsProcessor = PerpetualPositionsProcessor;
