// @klyra/cortex/dist/src/processors/account/subaccount.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubaccountProcessor = void 0;
const asset_positions_processor_1 = require("./asset-positions.processor");
const perpetual_positions_processor_1 = require("./perpetual-positions.processor");
const funding_payments_processor_1 = require("./funding-payments.processor");
const historical_pnls_processor_1 = require("./historical-pnls.processor");
const transfers_processor_1 = require("./transfers.processor");
const orders_processor_1 = require("./orders.processor");
const fills_processor_1 = require("./fills.processor");
class SubaccountProcessor {
    constructor(store) {
        this.store = store;
        this.assetPositionsProcessor = new asset_positions_processor_1.AssetPositionsProcessor(store);
        this.perpetualPositionsProcessor = new perpetual_positions_processor_1.PerpetualPositionsProcessor(store);
        this.fundingPaymentsProcessor = new funding_payments_processor_1.FundingPaymentsProcessor(store);
        this.historicalPnlsProcessor = new historical_pnls_processor_1.HistoricalPnlsProcessor(store);
        this.transfersProcessor = new transfers_processor_1.TransfersProcessor(store);
        this.ordersProcessor = new orders_processor_1.OrdersProcessor(store);
        this.fillsProcessor = new fills_processor_1.FillsProcessor(store);
    }
    process(msg, orders) {
        const existingUser = this.store.getState().user;
        const subaccountNumber = Number(msg.subaccountNumber);
        const existingSubaccount = existingUser.account.subaccounts[subaccountNumber];
        const newSubaccount = this.processSubaccount(existingSubaccount, msg, orders, true);
        return newSubaccount;
    }
    processUpdate(existingSubaccount, msg, subaccountNumber) {
        var _a, _b, _c, _d;
        const modified = existingSubaccount !== null && existingSubaccount !== void 0 ? existingSubaccount : {};
        for (const item of msg) {
            if (item.assetPositions) {
                const assetPositions = this.assetPositionsProcessor.process(item.assetPositions);
                if (assetPositions && assetPositions.length > 0) {
                    modified.assetPositions = assetPositions;
                }
            }
            if (item.perpetualPositions) {
                const perpetualPositions = this.perpetualPositionsProcessor.processChanges((_a = existingSubaccount === null || existingSubaccount === void 0 ? void 0 : existingSubaccount.positions) !== null && _a !== void 0 ? _a : {}, item.perpetualPositions);
                if (typeof perpetualPositions !== "undefined") {
                    modified.positions = perpetualPositions;
                    modified.openPositions = perpetualPositions;
                }
            }
            if (item.orders) {
                const orders = this.ordersProcessor.process((_b = existingSubaccount === null || existingSubaccount === void 0 ? void 0 : existingSubaccount.orders) !== null && _b !== void 0 ? _b : [], item.orders, subaccountNumber);
                if (orders.length > 0) {
                    modified.orders = orders;
                }
            }
            if (item.fills) {
                const fills = this.fillsProcessor.process((_c = existingSubaccount === null || existingSubaccount === void 0 ? void 0 : existingSubaccount.fills) !== null && _c !== void 0 ? _c : [], item.fills, subaccountNumber);
                if (fills.length > 0) {
                    modified.fills = fills;
                }
            }
            if (item.transfers) {
                const transfers = this.transfersProcessor.process((_d = existingSubaccount === null || existingSubaccount === void 0 ? void 0 : existingSubaccount.transfers) !== null && _d !== void 0 ? _d : [], item.transfers);
                if (transfers.length > 0) {
                    modified.transfers = transfers;
                }
            }
        }
        return modified;
    }
    processSubaccount(existing, payload, orders, firstTime) {
        var _a, _b;
        if (!payload) {
            return existing !== null && existing !== void 0 ? existing : null;
        }
        const modified = existing !== null && existing !== void 0 ? existing : {};
        const subaccountNumber = payload.subaccountNumber
            ? Number(payload.subaccountNumber)
            : 0;
        modified.subaccountNumber = subaccountNumber;
        modified.address = payload.address;
        modified.equity = Number(payload.equity);
        modified.freeCollateral = Number(payload.freeCollateral);
        modified.marginEnabled = payload.marginEnabled;
        if (firstTime) {
            const perpetualPositions = (_a = existing === null || existing === void 0 ? void 0 : existing.positions) !== null && _a !== void 0 ? _a : {};
            const newPerpetualPositions = this.perpetualPositionsProcessor.process(perpetualPositions, payload.openPerpetualPositions);
            modified.positions = newPerpetualPositions;
            modified.openPositions = newPerpetualPositions;
        }
        if (orders.length > 0) {
            const newOrders = this.ordersProcessor.process((_b = existing === null || existing === void 0 ? void 0 : existing.orders) !== null && _b !== void 0 ? _b : [], orders, subaccountNumber);
            modified.orders = newOrders;
        }
        if (Object.values(payload.assetPositions).length > 0) {
            const assetPositions = this.assetPositionsProcessor.process(Object.values(payload.assetPositions));
            if (assetPositions && assetPositions.length > 0) {
                modified.assetPositions = assetPositions;
            }
        }
        return modified;
    }
}
exports.SubaccountProcessor = SubaccountProcessor;
