// @klyra/cortex/dist/src/processors/account/transfers.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TransfersProcessor = void 0;
const merge_1 = require("../../utils/merge");
class TransfersProcessor {
    constructor(store) {
        this.store = store;
    }
    process(existing, payload) {
        if (!payload)
            return [];
        const newTransfers = payload.map((item) => {
            const id = item.transactionHash;
            return this.processTransfer(existing.find((transfer) => transfer.transactionHash === id), item);
        });
        return newTransfers;
    }
    processTransfer(existing, newItem) {
        // TODO: review types for transform
        const modified = (0, merge_1.transform)(existing, newItem, {
            string: {
                id: "id",
                type: "type",
                symbol: "asset",
                fromAddress: "fromAddress",
                toAddress: "toAddress",
                senderWallet: "fromAddress",
                recipientWallet: "toAddress",
                transactionHash: "transactionHash",
            },
            number: {
                size: "amount",
                createdAtHeight: "updatedAtBlock",
            },
        });
        const { sender, recipient } = newItem;
        modified.fromAddress = sender.address;
        modified.toAddress = recipient.address;
        if (!modified.id) {
            modified.id = newItem.transactionHash;
        }
        return modified;
    }
}
exports.TransfersProcessor = TransfersProcessor;
