// @klyra/cortex/dist/src/schemas/trade-forms.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateTradeForm = exports.tradeFormSchemas = void 0;
const zod_1 = require("zod");
const shared_1 = require("@klyra/shared");
const forms_1 = require("../types/forms");
const accounts_1 = require("../types/accounts");
const tradeFormBaseSchema = zod_1.z.object({
    type: zod_1.z.nativeEnum(forms_1.TradeFormType),
    marketId: zod_1.z.string().min(1, "Market ID is required"),
    side: zod_1.z.nativeEnum(shared_1.OrderSide),
    baseSize: zod_1.z.number().min(0, "Base size must be greater than 0"),
    quoteSize: zod_1.z.number().min(0, "Quote size must be greater than 0"),
    marginMode: zod_1.z.nativeEnum(accounts_1.MarginMode),
});
const marketTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.MARKET),
});
const limitTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.LIMIT),
    price: zod_1.z.number().min(0, "Price must be greater than 0"),
});
const stopLimitTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.STOP_LIMIT),
    price: zod_1.z.number().min(0, "Price must be greater than 0"),
    triggerPrice: zod_1.z.number().min(0, "Trigger price must be greater than 0"),
});
const takeProfitLimitTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.TAKE_PROFIT_LIMIT),
    price: zod_1.z.number().min(0, "Price must be greater than 0"),
    triggerPrice: zod_1.z.number().min(0, "Trigger price must be greater than 0"),
});
const stopMarketTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.STOP_MARKET),
    triggerPrice: zod_1.z.number().min(0, "Trigger price must be greater than 0"),
});
const takeProfitMarketTradeFormSchema = tradeFormBaseSchema.extend({
    type: zod_1.z.literal(forms_1.TradeFormType.TAKE_PROFIT_MARKET),
    triggerPrice: zod_1.z.number().min(0, "Trigger price must be greater than 0"),
});
exports.tradeFormSchemas = {
    [forms_1.TradeFormType.MARKET]: marketTradeFormSchema,
    [forms_1.TradeFormType.LIMIT]: limitTradeFormSchema,
    [forms_1.TradeFormType.STOP_LIMIT]: stopLimitTradeFormSchema,
    [forms_1.TradeFormType.TAKE_PROFIT_LIMIT]: takeProfitLimitTradeFormSchema,
    [forms_1.TradeFormType.STOP_MARKET]: stopMarketTradeFormSchema,
    [forms_1.TradeFormType.TAKE_PROFIT_MARKET]: takeProfitMarketTradeFormSchema,
};
const validateTradeForm = (form) => {
    const schema = exports.tradeFormSchemas[form.type];
    return schema.safeParse(form);
};
exports.validateTradeForm = validateTradeForm;
