// @klyra/core/dist/src/transactions/field-validations.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateFields = void 0;
const shared_1 = require("@klyra/shared");
const isDefined = (value) => value !== null && typeof value !== "undefined";
const isNumber = (value) => typeof value === "number" && !isNaN(value);
const isString = (value) => typeof value === "string" && value.length > 0;
const isBoolean = (value) => typeof value === "boolean";
const isEnum = (value, enumType) => Object.values(enumType).includes(value);
const fieldValidations = {
    subaccount: isDefined,
    ticker: isString,
    side: (side) => isEnum(side, shared_1.OrderSide),
    price: isNumber,
    triggerPrice: isNumber,
    size: isNumber,
    reduceOnly: isBoolean,
    goodTilBlock: isNumber,
    goodTilTime: isNumber,
    clientId: isNumber,
    recipientSubaccountNumber: isNumber,
    recipientAddress: isString,
    amount: isString,
};
const validateFields = (args) => {
    const responses = Object.entries(args).map(([key, value]) => fieldValidations[key](value));
    return responses.every((response) => response);
};
exports.validateFields = validateFields;
