// @klyra/shared/dist/types/orders.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OrderExecution = exports.OrderTimeInForce = exports.OrderSide = exports.OrderType = exports.OrderStatusEnum = void 0;
var OrderStatusEnum;
(function (OrderStatusEnum) {
    OrderStatusEnum["CANCELLED"] = "CANCELLED";
    OrderStatusEnum["CANCELLING"] = "CANCELLING";
    OrderStatusEnum["FILLED"] = "FILLED";
    OrderStatusEnum["OPEN"] = "OPEN";
    OrderStatusEnum["PENDING"] = "PENDING";
    OrderStatusEnum["UNTRIGGERED"] = "UNTRIGGERED";
    OrderStatusEnum["PARTIALLY_FILLED"] = "PARTIALLY_FILLED";
    OrderStatusEnum["PARTIALLY_CANCELLED"] = "PARTIALLY_CANCELLED";
})(OrderStatusEnum || (exports.OrderStatusEnum = OrderStatusEnum = {}));
var OrderType;
(function (OrderType) {
    OrderType["LIMIT"] = "LIMIT";
    OrderType["MARKET"] = "MARKET";
    OrderType["STOP_LIMIT"] = "STOP_LIMIT";
    OrderType["STOP_MARKET"] = "STOP_MARKET";
    OrderType["TAKE_PROFIT_LIMIT"] = "TAKE_PROFIT_LIMIT";
    OrderType["TAKE_PROFIT_MARKET"] = "TAKE_PROFIT_MARKET";
    OrderType["TRAILING_STOP"] = "TRAILING_STOP";
    OrderType["LIQUIDATED"] = "LIQUIDATED";
    OrderType["LIQUIDATION"] = "LIQUIDATION";
    OrderType["OFFSETTING"] = "OFFSETING";
    OrderType["DELEVERAGED"] = "DELEVERAGED";
    OrderType["FINAL_SETTLEMENT"] = "FINAL_SETTLEMENT";
})(OrderType || (exports.OrderType = OrderType = {}));
var OrderSide;
(function (OrderSide) {
    OrderSide["BUY"] = "BUY";
    OrderSide["SELL"] = "SELL";
})(OrderSide || (exports.OrderSide = OrderSide = {}));
var OrderTimeInForce;
(function (OrderTimeInForce) {
    OrderTimeInForce["GTT"] = "GTT";
    OrderTimeInForce["IOC"] = "IOC";
})(OrderTimeInForce || (exports.OrderTimeInForce = OrderTimeInForce = {}));
var OrderExecution;
(function (OrderExecution) {
    OrderExecution["DEFAULT"] = "DEFAULT";
    OrderExecution["IOC"] = "IOC";
    OrderExecution["POST_ONLY"] = "POST_ONLY";
})(OrderExecution || (exports.OrderExecution = OrderExecution = {}));
