// @klyra/cortex/src/__tests__/mocks/accounts.mock.ts
import {
  IndexerOrderStatus,
  IndexerPositionSide,
  type IndexerSubaccountBatchUpdateMessage,
  type IndexerSubaccountSubscribedMessage,
  MessageChannel,
  MessageType,
} from "@klyra/shared";

export const SUBSCRIBED_MESSAGE = {
  type: MessageType.SUBSCRIBED,
  connection_id: "91459969-3e05-487d-94dd-c8b59ee05a58",
  message_id: 2,
  channel: MessageChannel.PARENT_SUBACCOUNTS,
  id: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn/0",
  contents: {
    subaccount: {
      address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
      parentSubaccountNumber: 0,
      equity: "415.171879998",
      freeCollateral: "314.9069364984",
      childSubaccounts: [
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 0,
          equity: "415.171879998",
          freeCollateral: "314.9069364984",
          openPerpetualPositions: {
            "RENDER-USD": {
              market: "RENDER-USD",
              status: IndexerOrderStatus.OPEN,
              side: IndexerPositionSide.SHORT,
              size: "-0.01",
              maxSize: "-0.0094",
              entryPrice: "57925.9",
              exitPrice: null,
              realizedPnl: "0",
              unrealizedPnl: "-56.388",
              createdAt: "2024-09-16T22:51:13.140Z",
              createdAtHeight: "21330630",
              closedAt: null,
              sumOpen: "0.01",
              sumClose: "0",
              netFunding: "0",
              subaccountNumber: 0,
            },
            "BTC-USD": {
              market: "BTC-USD",
              status: IndexerOrderStatus.OPEN,
              side: IndexerPositionSide.SHORT,
              size: "-0.01",
              maxSize: "-0.0094",
              entryPrice: "57925.9",
              exitPrice: null,
              realizedPnl: "0",
              unrealizedPnl: "-56.388",
              createdAt: "2024-09-16T22:51:13.140Z",
              createdAtHeight: "21330630",
              closedAt: null,
              sumOpen: "0.01",
              sumClose: "0",
              netFunding: "0",
              subaccountNumber: 0,
            },
            "ETH-USD": {
              market: "ETH-USD",
              status: IndexerOrderStatus.OPEN,
              side: IndexerPositionSide.LONG,
              size: "0.501",
              maxSize: "1.039",
              entryPrice: "2369.96920077972709551657",
              exitPrice: "2346.92803468208092485549",
              realizedPnl: "-23.91673040935672514620104",
              unrealizedPnl: "132.91570040935672514619843",
              createdAt: "2024-09-12T18:05:15.725Z",
              createdAtHeight: "21019737",
              closedAt: null,
              sumOpen: "1.539",
              sumClose: "1.038",
              netFunding: "0",
              subaccountNumber: 0,
            },
          },
          assetPositions: {
            USDC: {
              size: "281.79679",
              symbol: "USDC",
              side: IndexerPositionSide.SHORT,
              assetId: "0",
              subaccountNumber: 0,
            },
          },
          marginEnabled: true,
          updatedAtHeight: "21539244",
          latestProcessedBlockHeight: "21585658",
        },
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 128,
          equity: "0",
          freeCollateral: "0",
          openPerpetualPositions: {},
          assetPositions: {},
          marginEnabled: true,
          updatedAtHeight: "15432184",
          latestProcessedBlockHeight: "21585658",
        },
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 256,
          equity: "0",
          freeCollateral: "0",
          openPerpetualPositions: {},
          assetPositions: {},
          marginEnabled: true,
          updatedAtHeight: "15406589",
          latestProcessedBlockHeight: "21585658",
        },
      ],
    },
    orders: [],
    blockHeight: "21585658",
  },
} as unknown as IndexerSubaccountSubscribedMessage;

export const BATCH_UPDATE_MESSAGE = {
  type: "channel_batch_data",
  connection_id: "dec967d8-dd45-4811-ba17-bc5dcba822ff",
  message_id: 919,
  id: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn/0",
  channel: "v4_parent_subaccounts",
  version: "3.0.0",
  contents: [
    {
      perpetualPositions: [
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 0,
          positionId: "02073363-b418-58e4-a83a-aa06d686ced6",
          market: "RENDER-USD",
          side: "LONG",
          status: "CLOSED",
          size: "0",
          maxSize: "2",
          netFunding: "-0.106652",
          entryPrice: "5.069",
          exitPrice: null,
          sumOpen: "2",
          sumClose: "0",
          realizedPnl: "-0.106652",
          unrealizedPnl: "0",
        },
      ],
      assetPositions: [
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 0,
          positionId: "3507ea17-7a70-5213-8afe-a1e1603719c7",
          assetId: "0",
          symbol: "USDC",
          side: "LONG",
          size: "633.540591",
        },
      ],
      blockHeight: "21699529",
    },
    {
      fills: [
        {
          id: "85713633-0985-52b9-a4dd-fb6661059ad4",
          fee: "0.005982",
          side: "SELL",
          size: "2",
          type: "LIMIT",
          price: "5.982",
          eventId: "014b1bc90000000200000002",
          orderId: "f0169af7-acde-5fb3-9105-cb9a08cddbb2",
          createdAt: "2024-09-26T03:13:37.577Z",
          liquidity: "TAKER",
          clobPairId: "65",
          quoteAmount: "11.964",
          subaccountId: "527eae56-8ec3-5f01-ada1-1372ad50b36b",
          clientMetadata: "1",
          createdAtHeight: "21699529",
          transactionHash:
            "9BCFA3067AACF7835EB8940B26898DE117F2DA6CC62D4D4AD8B186B025547877",
          ticker: "RENDER-USD",
        },
      ],
      perpetualPositions: [
        {
          address: "klyra1qgjmxkz09m495dackm88r2kghwepawurvsjtrn",
          subaccountNumber: 0,
          positionId: "02073363-b418-58e4-a83a-aa06d686ced6",
          market: "RENDER-USD",
          side: "LONG",
          status: "CLOSED",
          size: "0",
          maxSize: "2",
          netFunding: "-0.106652",
          entryPrice: "5.069",
          exitPrice: "5.982",
          sumOpen: "2",
          sumClose: "2",
          realizedPnl: "1.719348",
          unrealizedPnl: "0",
        },
      ],
      blockHeight: "21699529",
      orders: [
        {
          id: "f0169af7-acde-5fb3-9105-cb9a08cddbb2",
          side: "SELL",
          size: "2",
          type: "LIMIT",
          price: "0.001",
          status: "FILLED",
          clientId: "1337949897",
          updatedAt: "2024-09-26T03:13:37.577Z",
          clobPairId: "65",
          orderFlags: "0",
          reduceOnly: true,
          timeInForce: "IOC",
          totalFilled: "2",
          goodTilBlock: "21699537",
          subaccountId: "527eae56-8ec3-5f01-ada1-1372ad50b36b",
          triggerPrice: null,
          clientMetadata: "1",
          createdAtHeight: "21699529",
          updatedAtHeight: "21699529",
          goodTilBlockTime: null,
          postOnly: false,
          ticker: "RENDER-USD",
        },
      ],
    },
    {
      tradingReward: {
        tradingReward: "0.000961090583950518",
        createdAtHeight: "21699529",
        createdAt: "2024-09-26T03:13:37.577Z",
      },
      blockHeight: "21699529",
    },
  ],
  subaccountNumber: 0,
} as unknown as IndexerSubaccountBatchUpdateMessage;
