// @klyra/cortex/src/__tests__/mocks/markets.mock.ts
import {
  type IndexerMarketSubscribedMessage,
  type IndexerMarketBatchUpdateMessage,
  MessageChannel,
  MessageType,
} from "@klyra/shared";

export const SUBSCRIBED_MESSAGE = {
  type: MessageType.SUBSCRIBED,
  connection_id: "c4773542-f710-4069-8211-83840c6b010d",
  message_id: 1,
  channel: MessageChannel.MARKETS,
  contents: {
    markets: {
      "BTC-USD": {
        clobPairId: "0",
        ticker: "BTC-USD",
        status: "ACTIVE",
        spotPrice: "63360.41819",
        priceChange24H: "-395.10181",
        volume24H: "10860.1122",
        trades24H: 1663,
        nextFundingRate: "0",
        initialMarginFraction: "0.05",
        maintenanceMarginFraction: "0.03",
        openInterest: "50.3349",
        atomicResolution: -10,
        quantumConversionExponent: -9,
        tickSize: "1",
        stepSize: "0.0001",
        stepBaseQuantums: 1000000,
        subticksPerTick: 100000,
        marketType: "CROSS",
        openInterestLowerCap: "0",
        openInterestUpperCap: "0",
        baseOpenInterest: "50.3349",
      },
      "ETH-USD": {
        clobPairId: "1",
        ticker: "ETH-USD",
        status: "ACTIVE",
        spotPrice: "2582.364261",
        priceChange24H: "-55.325739",
        volume24H: "21223.3024",
        trades24H: 2109,
        nextFundingRate: "0",
        initialMarginFraction: "0.05",
        maintenanceMarginFraction: "0.03",
        openInterest: "964.107",
        atomicResolution: -9,
        quantumConversionExponent: -9,
        tickSize: "0.1",
        stepSize: "0.001",
        stepBaseQuantums: 1000000,
        subticksPerTick: 100000,
        marketType: "CROSS",
        openInterestLowerCap: "0",
        openInterestUpperCap: "0",
        baseOpenInterest: "964.107",
      },
    },
  },
} as unknown as IndexerMarketSubscribedMessage;

export const ORACLE_PRICE_UPDATE_MESSAGE = {
  type: MessageType.CHANNEL_BATCH_DATA,
  connection_id: "c4773542-f710-4069-8211-83840c6b010d",
  message_id: 214,
  channel: MessageChannel.MARKETS,
  version: "1.0.0",
  contents: [
    {
      oraclePrices: {
        "BTC-USD": { id: "0", spotPrice: "63360.41819" },
        "ETH-USD": { id: "1", spotPrice: "2582.364261" },
      },
    },
  ],
} as unknown as IndexerMarketBatchUpdateMessage;

export const TRADING_MARKET_UPDATE_MESSAGE = {
  type: MessageType.CHANNEL_BATCH_DATA,
  connection_id: "c4773542-f710-4069-8211-83840c6b010d",
  message_id: 214,
  channel: MessageChannel.MARKETS,
  version: "1.0.0",
  contents: [
    {
      trading: {
        "BTC-USD": {
          id: "0",
          trades24H: 10000,
        },
      },
    },
  ],
} as unknown as IndexerMarketBatchUpdateMessage;
