// @klyra/cortex/src/__tests__/mocks/trades.mock.ts
import {
  type IndexerTradeBatchUpdateMessage,
  type IndexerTradeSubscribedMessage,
  MessageChannel,
  MessageType,
} from "@klyra/shared";

export const SUBSCRIBED_MESSAGE = {
  type: MessageType.SUBSCRIBED,
  connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
  message_id: 3,
  channel: MessageChannel.TRADES,
  id: "ETH-USD",
  contents: {
    trades: [
      {
        id: "014a8f900000000200000014",
        side: "BUY",
        size: "0.002",
        price: "2632.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:45:46.193Z",
        createdAtHeight: "21663632",
      },
      {
        id: "014a8f900000000200000005",
        side: "BUY",
        size: "0.002",
        price: "2632.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:45:46.193Z",
        createdAtHeight: "21663632",
      },
      {
        id: "014a8f60000000020000000b",
        side: "BUY",
        size: "0.002",
        price: "2632.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:44:51.462Z",
        createdAtHeight: "21663584",
      },
      {
        id: "014a8f42000000020000000e",
        side: "BUY",
        size: "0.002",
        price: "2632.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:44:13.745Z",
        createdAtHeight: "21663554",
      },
      {
        id: "014a8f0e000000020000000b",
        side: "BUY",
        size: "0.002",
        price: "2634.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:43:13.564Z",
        createdAtHeight: "21663502",
      },
      {
        id: "014a8f0e0000000200000008",
        side: "BUY",
        size: "0.002",
        price: "2634.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:43:13.564Z",
        createdAtHeight: "21663502",
      },
      {
        id: "014a8f0e0000000200000005",
        side: "BUY",
        size: "0.002",
        price: "2634.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:43:13.564Z",
        createdAtHeight: "21663502",
      },
      {
        id: "014a8f0e0000000200000002",
        side: "BUY",
        size: "0.002",
        price: "2634.6",
        type: "LIMIT",
        createdAt: "2024-09-25T14:43:13.564Z",
        createdAtHeight: "21663502",
      },
      {
        id: "014a8eff000000020000000b",
        side: "BUY",
        size: "0.002",
        price: "2634.5",
        type: "LIMIT",
        createdAt: "2024-09-25T14:42:56.767Z",
        createdAtHeight: "21663487",
      },
      {
        id: "014a8eff0000000200000008",
        side: "BUY",
        size: "0.002",
        price: "2634.5",
        type: "LIMIT",
        createdAt: "2024-09-25T14:42:56.767Z",
        createdAtHeight: "21663487",
      },
    ],
  },
} as unknown as IndexerTradeSubscribedMessage;

export const BATCH_UPDATE_MESSAGE = {
  type: MessageType.CHANNEL_BATCH_DATA,
  connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
  message_id: 2449,
  id: "ETH-USD",
  channel: MessageChannel.TRADES,
  version: "2.1.0",
  contents: [
    {
      trades: [
        {
          id: "014a968f0000000200000005",
          size: "0.002",
          price: "2634.1",
          side: "BUY",
          createdAt: "2024-09-25T15:22:29.031Z",
          type: "LIMIT",
        },
      ],
    },
  ],
} as unknown as IndexerTradeBatchUpdateMessage;
