// @klyra/ws/dist/manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketManager = void 0;
const ws_1 = require("ws");
const shared_1 = require("@klyra/shared");
const shared_2 = require("@klyra/shared");
class WebSocketManager {
    constructor(config) {
        this.ws = null;
        this.intentionalClose = false;
        this.isConnected = false;
        this.subscribedChannels = [];
        this.url = config.url;
        this.reconnectInterval = config.reconnectInterval || 5000;
        this.eventEmitter = new shared_2.EventEmitter();
    }
    // checks if we are on the client or server and returns the appropriate WebSocket
    getWebSocket() {
        if (typeof window !== "undefined") {
            return new WebSocket(this.url);
        }
        else {
            return new ws_1.WebSocket(this.url);
        }
    }
    connect() {
        this.intentionalClose = false;
        this.ws = this.getWebSocket();
        this.ws.onopen = () => {
            console.log("WebSocket is now open!");
            this.isConnected = true;
        };
        this.ws.onmessage = (event) => {
            console.log("NEW WS MSG", event.data.toString());
            try {
                const message = JSON.parse(event.data.toString());
                // add to susbcribed messages if it is a subscribed channel
                if (message.type === shared_1.MessageType.SUBSCRIBE) {
                    this.subscribedChannels.push([message.channel, message.id]);
                }
                // if it's a unsubscribe message, remove from subscribed channels
                if (message.type === shared_1.MessageType.UNSUBSCRIBED) {
                    this.subscribedChannels = this.subscribedChannels.filter(([channel, id]) => channel !== message.channel || (id && id !== message.id));
                }
                this.eventEmitter.emit("message", message);
            }
            catch (error) {
                console.error("Error parsing WebSocket message:", error);
            }
        };
        this.ws.onclose = () => {
            console.log("WebSocket disconnected.");
            this.eventEmitter.emit("disconnected");
            this.isConnected = false;
            if (!this.intentionalClose) {
                setTimeout(() => this.connect(), this.reconnectInterval);
            }
        };
        this.ws.onerror = (error) => {
            console.error("WebSocket error:", error);
            this.eventEmitter.emit("error", error);
        };
    }
    send(message) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            // if it's a subscribe message, and the channel is already subscribed, don't send it
            if (message.type === shared_1.MessageType.SUBSCRIBE) {
                const isSubscribed = this.subscribedChannels.some(([channel, id]) => channel === message.channel && id && id === message.id);
                if (isSubscribed) {
                    return;
                }
            }
            this.ws.send(JSON.stringify(message));
        }
        else {
            console.error("WebSocket is not open");
        }
    }
    on(event, callback) {
        this.eventEmitter.on(event, callback);
    }
    off(event, callback) {
        this.eventEmitter.off(event, callback);
    }
    disconnect() {
        var _a;
        this.intentionalClose = true;
        (_a = this.ws) === null || _a === void 0 ? void 0 : _a.close();
        this.subscribedChannels = [];
    }
}
exports.WebSocketManager = WebSocketManager;
