// @klyra/cortex/dist/src/types/accounts.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PositionSide = exports.FillLiquidity = exports.MarginMode = void 0;
var MarginMode;
(function (MarginMode) {
    MarginMode["CROSS"] = "CROSS";
    MarginMode["ISOLATED"] = "ISOLATED";
})(MarginMode || (exports.MarginMode = MarginMode = {}));
var FillLiquidity;
(function (FillLiquidity) {
    FillLiquidity["MAKER"] = "MAKER";
    FillLiquidity["TAKER"] = "TAKER";
})(FillLiquidity || (exports.FillLiquidity = FillLiquidity = {}));
var PositionSide;
(function (PositionSide) {
    PositionSide["LONG"] = "LONG";
    PositionSide["SHORT"] = "SHORT";
    PositionSide["NONE"] = "NONE";
})(PositionSide || (exports.PositionSide = PositionSide = {}));
var TransferRecordType;
(function (TransferRecordType) {
    TransferRecordType["DEPOSIT"] = "DEPOSIT";
    TransferRecordType["WITHDRAW"] = "WITHDRAW";
    TransferRecordType["TRANSFER_IN"] = "TRANSFER_IN";
    TransferRecordType["TRANSFER_OUT"] = "TRANSFER_OUT";
})(TransferRecordType || (TransferRecordType = {}));
