// @klyra/cortex/dist/src/types/configs.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CandlesPeriod = exports.HistoricalPnlPeriod = exports.ApiStatus = exports.NetworkStatus = void 0;
var NetworkStatus;
(function (NetworkStatus) {
    NetworkStatus["UNKNOWN"] = "UNKNOWN";
    NetworkStatus["UNREACHABLE"] = "UNREACHABLE";
    NetworkStatus["HALTED"] = "HALTED";
    NetworkStatus["NORMAL"] = "NORMAL";
})(NetworkStatus || (exports.NetworkStatus = NetworkStatus = {}));
var ApiStatus;
(function (ApiStatus) {
    ApiStatus["UNKNOWN"] = "UNKNOWN";
    ApiStatus["VALIDATOR_DOWN"] = "VALIDATOR_DOWN";
    ApiStatus["VALIDATOR_HALTED"] = "VALIDATOR_HALTED";
    ApiStatus["INDEXER_DOWN"] = "INDEXER_DOWN";
    ApiStatus["INDEXER_HALTED"] = "INDEXER_HALTED";
    ApiStatus["INDEXER_TRAILING"] = "INDEXER_TRAILING";
    ApiStatus["NORMAL"] = "NORMAL";
})(ApiStatus || (exports.ApiStatus = ApiStatus = {}));
var HistoricalPnlPeriod;
(function (HistoricalPnlPeriod) {
    HistoricalPnlPeriod["ONE_DAY"] = "1D";
    HistoricalPnlPeriod["SEVEN_DAYS"] = "7D";
    HistoricalPnlPeriod["THIRTY_DAYS"] = "30D";
    HistoricalPnlPeriod["NINETY_DAYS"] = "90D";
})(HistoricalPnlPeriod || (exports.HistoricalPnlPeriod = HistoricalPnlPeriod = {}));
var CandlesPeriod;
(function (CandlesPeriod) {
    CandlesPeriod["ONE_MINUTE"] = "1M";
    CandlesPeriod["FIVE_MINUTES"] = "5M";
    CandlesPeriod["FIFTEEN_MINUTES"] = "15M";
    CandlesPeriod["THIRTY_MINUTES"] = "30M";
    CandlesPeriod["ONE_HOUR"] = "1H";
    CandlesPeriod["FOUR_HOURS"] = "4H";
    CandlesPeriod["ONE_DAY"] = "1D";
})(CandlesPeriod || (exports.CandlesPeriod = CandlesPeriod = {}));
