// @klyra/shared/dist/types/indexer/account.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerOrderType = exports.IndexerTransferType = exports.IndexerLiquidity = exports.IndexerFillType = exports.IndexerOrderStatus = exports.IndexerTimeInForce = exports.IndexerPerpetualPositionStatus = exports.IndexerOrderSide = exports.IndexerPositionSide = void 0;
var IndexerPositionSide;
(function (IndexerPositionSide) {
    IndexerPositionSide["LONG"] = "LONG";
    IndexerPositionSide["SHORT"] = "SHORT";
})(IndexerPositionSide || (exports.IndexerPositionSide = IndexerPositionSide = {}));
var IndexerOrderSide;
(function (IndexerOrderSide) {
    IndexerOrderSide["BUY"] = "BUY";
    IndexerOrderSide["SELL"] = "SELL";
})(IndexerOrderSide || (exports.IndexerOrderSide = IndexerOrderSide = {}));
var IndexerPerpetualPositionStatus;
(function (IndexerPerpetualPositionStatus) {
    IndexerPerpetualPositionStatus["OPEN"] = "OPEN";
    IndexerPerpetualPositionStatus["CLOSED"] = "CLOSED";
    IndexerPerpetualPositionStatus["LIQUIDATED"] = "LIQUIDATED";
})(IndexerPerpetualPositionStatus || (exports.IndexerPerpetualPositionStatus = IndexerPerpetualPositionStatus = {}));
var IndexerTimeInForce;
(function (IndexerTimeInForce) {
    IndexerTimeInForce["GTT"] = "GTT";
    IndexerTimeInForce["FOK"] = "FOK";
    IndexerTimeInForce["IOC"] = "IOC";
    IndexerTimeInForce["POST_ONLY"] = "POST_ONLY";
})(IndexerTimeInForce || (exports.IndexerTimeInForce = IndexerTimeInForce = {}));
var IndexerOrderStatus;
(function (IndexerOrderStatus) {
    IndexerOrderStatus["OPEN"] = "OPEN";
    IndexerOrderStatus["FILLED"] = "FILLED";
    IndexerOrderStatus["CANCELED"] = "CANCELED";
    IndexerOrderStatus["BEST_EFFORT_CANCELED"] = "BEST_EFFORT_CANCELED";
    IndexerOrderStatus["BEST_EFFORT_OPENED"] = "BEST_EFFORT_OPENED";
    IndexerOrderStatus["UNTRIGGERED"] = "UNTRIGGERED";
})(IndexerOrderStatus || (exports.IndexerOrderStatus = IndexerOrderStatus = {}));
var IndexerFillType;
(function (IndexerFillType) {
    IndexerFillType["LIMIT"] = "LIMIT";
    IndexerFillType["LIQUIDATED"] = "LIQUIDATED";
    IndexerFillType["LIQUIDATION"] = "LIQUIDATION";
    IndexerFillType["DELEVERAGED"] = "DELEVERAGED";
    IndexerFillType["OFFSETTING"] = "OFFSETTING";
})(IndexerFillType || (exports.IndexerFillType = IndexerFillType = {}));
var IndexerLiquidity;
(function (IndexerLiquidity) {
    IndexerLiquidity["TAKER"] = "TAKER";
    IndexerLiquidity["MAKER"] = "MAKER";
})(IndexerLiquidity || (exports.IndexerLiquidity = IndexerLiquidity = {}));
var IndexerTransferType;
(function (IndexerTransferType) {
    IndexerTransferType["TRANSFER_IN"] = "TRANSFER_IN";
    IndexerTransferType["TRANSFER_OUT"] = "TRANSFER_OUT";
    IndexerTransferType["DEPOSIT"] = "DEPOSIT";
    IndexerTransferType["WITHDRAWAL"] = "WITHDRAWAL";
})(IndexerTransferType || (exports.IndexerTransferType = IndexerTransferType = {}));
var IndexerOrderType;
(function (IndexerOrderType) {
    IndexerOrderType["LIMIT"] = "LIMIT";
    IndexerOrderType["MARKET"] = "MARKET";
    IndexerOrderType["STOP_LIMIT"] = "STOP_LIMIT";
    IndexerOrderType["STOP_MARKET"] = "STOP_MARKET";
    IndexerOrderType["TRAILING_STOP"] = "TRAILING_STOP";
    IndexerOrderType["TAKE_PROFIT"] = "TAKE_PROFIT";
    IndexerOrderType["TAKE_PROFIT_MARKET"] = "TAKE_PROFIT_MARKET";
    IndexerOrderType["HARD_TRADE"] = "HARD_TRADE";
    IndexerOrderType["FAILED_HARD_TRADE"] = "FAILED_HARD_TRADE";
    IndexerOrderType["TRANSFER_PLACEHOLDER"] = "TRANSFER_PLACEHOLDER";
})(IndexerOrderType || (exports.IndexerOrderType = IndexerOrderType = {}));
