// @klyra/shared/dist/types/indexer/markets.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.IndexerPerpetualMarketType = exports.IndexerPerpetualMarketStatus = void 0;
var IndexerPerpetualMarketStatus;
(function (IndexerPerpetualMarketStatus) {
    IndexerPerpetualMarketStatus["ACTIVE"] = "ACTIVE";
    IndexerPerpetualMarketStatus["PAUSED"] = "PAUSED";
    IndexerPerpetualMarketStatus["CANCEL_ONLY"] = "CANCEL_ONLY";
    IndexerPerpetualMarketStatus["POST_ONLY"] = "POST_ONLY";
    IndexerPerpetualMarketStatus["INITIALIZING"] = "INITIALIZING";
    IndexerPerpetualMarketStatus["FINAL_SETTLEMENT"] = "FINAL_SETTLEMENT";
})(IndexerPerpetualMarketStatus || (exports.IndexerPerpetualMarketStatus = IndexerPerpetualMarketStatus = {}));
var IndexerPerpetualMarketType;
(function (IndexerPerpetualMarketType) {
    IndexerPerpetualMarketType["CROSS"] = "CROSS";
    IndexerPerpetualMarketType["ISOLATED"] = "ISOLATED";
})(IndexerPerpetualMarketType || (exports.IndexerPerpetualMarketType = IndexerPerpetualMarketType = {}));
