// @klyra/cortex/dist/src/processors/forms/trade/trade-form-summary.calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradeFormSummaryCalculator = void 0;
const shared_1 = require("@klyra/shared");
const types_1 = require("../../../types");
const orderbook_1 = require("../../../utils/orderbook");
const indexer_conversion_1 = require("../../../utils/indexer-conversion");
class TradeFormSummaryCalculator {
    constructor(store) {
        this.store = store;
    }
    calculate(calculationResult, subaccount) {
        var _a, _b, _c, _d;
        const multiplier = calculationResult.side === shared_1.OrderSide.BUY ? 1 : -1;
        const market = this.store.getState().markets[calculationResult.marketId];
        if (!market)
            return null;
        switch (calculationResult.type) {
            case types_1.TradeFormType.MARKET: {
                // TODO: get fee rate from user
                const feeRate = 0;
                const values = calculationResult;
                // const midMarketPrice = this.orderbookMidPrice(
                //   calculationResult.marketId
                // );
                // const worstPrice = calculationResult.worstPrice;
                // const slippageFromMidPrice = ...
                const price = (_b = (_a = values.internal) === null || _a === void 0 ? void 0 : _a.price) !== null && _b !== void 0 ? _b : 0;
                const side = values.side;
                // TODO: set custom slippage
                const maxSlippage = 0.01;
                let payloadPrice = null;
                if (price) {
                    payloadPrice =
                        side === shared_1.OrderSide.BUY
                            ? price * (1 + maxSlippage)
                            : price * (1 - maxSlippage);
                }
                const quoteSize = values.quoteSize;
                const baseSize = price && quoteSize ? price * quoteSize : null;
                const fee = baseSize && feeRate ? baseSize * feeRate : null;
                // TODO: check if the -1 is correct
                const total = baseSize ? baseSize * multiplier + (fee !== null && fee !== void 0 ? fee : 0) * -1 : null;
                const spotPrice = market.spotPrice;
                const priceDiff = this.slippage(price, spotPrice, side);
                // TODO: check if we should round here (abacus does, but if we do it breaks the number)
                // const indexSlippage =
                //   priceDiff && spotPrice ?
                //      Rounder.quickRound(priceDiff / spotPrice, SLIPPAGE_STEP_SIZE)
                //     : null;
                const indexSlippage = priceDiff && spotPrice ? priceDiff / spotPrice : null;
                const filled = (_d = (_c = values.internal) === null || _c === void 0 ? void 0 : _c.filled) !== null && _d !== void 0 ? _d : false;
                // TODO: These are the current values. Ideally, we would calculate the new values and send both to the frontend.
                const positionMargin = this.getPositionMargin(values.marginMode, market, subaccount);
                // const positionLeverage = this.getPositionLeverage(market, subaccount);
                // TODO: calculate new account equity
                return {
                    price,
                    payloadPrice,
                    baseSize,
                    quoteSize,
                    fee,
                    feeRate,
                    total,
                    // slippage: slippageFromMidPrice,
                    filled,
                    indexSlippage,
                    positionMargin,
                };
            }
            case types_1.TradeFormType.LIMIT: {
                const values = calculationResult;
                // const timeInForce = values.timeInForce;
                // const execution = values.execution;
                // const isMaker =
                //   timeInForce === OrderTimeInForce.GTT ||
                //   execution === OrderExecution.POST_ONLY;
                // TODO: get fee rate from user
                const feeRate = 0;
                const price = values.price;
                const quoteSize = values.quoteSize;
                const baseSize = price && quoteSize ? price * quoteSize : null;
                const fee = baseSize && feeRate ? baseSize * feeRate : null;
                const total = baseSize ? baseSize * multiplier + (fee !== null && fee !== void 0 ? fee : 0) * -1 : null;
                const positionMargin = this.getPositionMargin(values.marginMode, market, subaccount);
                // const positionLeverage = this.getPositionLeverage(market, subaccount);
                // TODO: calculate new account equity
                return {
                    price,
                    payloadPrice: price,
                    baseSize,
                    quoteSize,
                    fee,
                    feeRate,
                    total,
                    filled: true,
                    positionMargin,
                };
            }
            default:
                return null;
        }
    }
    orderbookMidPrice(marketId) {
        const orderbook = (0, orderbook_1.getOrderbookFromStore)(this.store, marketId);
        const asks = orderbook.asks;
        const bids = orderbook.bids;
        const firstAsk = asks[0];
        const firstBid = bids[0];
        if (!firstAsk || !firstBid)
            return null;
        const midPrice = (firstAsk.price + firstBid.price) / 2;
        return midPrice;
    }
    slippage(price, spotPrice, side) {
        if (!price || !spotPrice)
            return null;
        return side === shared_1.OrderSide.BUY ? price - spotPrice : spotPrice - price;
    }
    getPositionMargin(marginMode, market, subaccount) {
        var _a;
        const position = (_a = subaccount.openPositions) === null || _a === void 0 ? void 0 : _a[market.id];
        if (!position)
            return null;
        if (marginMode === types_1.MarginMode.CROSS) {
            const currentNotionalTotal = position.notionalTotal;
            const mmf = market.configs.maintenanceMarginFraction;
            if (mmf && currentNotionalTotal) {
                return currentNotionalTotal * mmf;
            }
        }
        else {
            const currentEquity = position.equity;
            if (currentEquity) {
                return currentEquity;
            }
        }
        return null;
    }
    getPositionLeverage(market, subaccount) {
        const { subaccountNumber, leverage, openPositions } = subaccount;
        const marginMode = (0, indexer_conversion_1.getMarginModeFromSubaccount)(subaccountNumber);
        if (marginMode === types_1.MarginMode.ISOLATED) {
            return leverage !== null && leverage !== void 0 ? leverage : null;
        }
        const marketId = market.id;
        const position = openPositions === null || openPositions === void 0 ? void 0 : openPositions[marketId];
        const currentLeverage = position === null || position === void 0 ? void 0 : position.leverage;
        return currentLeverage !== null && currentLeverage !== void 0 ? currentLeverage : null;
    }
    getExecutionPrice(spotPrice, limitPrice, isBuying) {
        if (!spotPrice || !limitPrice)
            return null;
        if (isBuying) {
            return Math.min(spotPrice, limitPrice);
        }
        return Math.max(spotPrice, limitPrice);
    }
}
exports.TradeFormSummaryCalculator = TradeFormSummaryCalculator;
