// @klyra/cortex/dist/src/processors/forms/trade/trade-form-options.calculator.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradeFormOptionsCalculator = void 0;
const shared_1 = require("@klyra/shared");
const types_1 = require("../../../types");
const values_1 = require("./values");
class TradeFormOptionsCalculator {
    constructor(store) {
        this.store = store;
    }
    calculate(formInputs) {
        const formType = formInputs.type;
        const typeToFnMap = {
            [types_1.TradeFormType.MARKET]: this.getMarketTypeFormOptions.bind(this),
            [types_1.TradeFormType.LIMIT]: this.getLimitTypeFormOptions.bind(this),
            [types_1.TradeFormType.STOP_LIMIT]: this.getSLTPLimitTypeFormOptions.bind(this),
            [types_1.TradeFormType.TAKE_PROFIT_LIMIT]: this.getSLTPLimitTypeFormOptions.bind(this),
            [types_1.TradeFormType.STOP_MARKET]: this.getSLTPMarketTypeFormOptions.bind(this),
            [types_1.TradeFormType.TAKE_PROFIT_MARKET]: this.getSLTPMarketTypeFormOptions.bind(this),
        };
        const options = typeToFnMap[formType](formInputs);
        options.needsMarginMode = this.getNeedsMarginMode(formInputs);
        const newOptions = { ...values_1.DEFAULT_TRADE_FORM_OPTIONS, ...options };
        return newOptions;
    }
    getMarketTypeFormOptions(_) {
        const options = {};
        options.needsReduceOnly = true;
        options.needsGoodTilTime = true;
        options.needsExecution = true;
        return options;
    }
    getLimitTypeFormOptions(formInputs) {
        const options = {};
        const isGoodTilTime = formInputs.timeInForce === shared_1.OrderTimeInForce.GTT;
        options.needsPrice = true;
        options.needsTimeInForce = true;
        options.needsGoodTilTime = isGoodTilTime;
        options.needsPostOnly = isGoodTilTime;
        options.needsReduceOnly = !isGoodTilTime;
        return options;
    }
    getSLTPMarketTypeFormOptions(_) {
        const options = {};
        options.needsTriggerPrice = true;
        options.needsGoodTilTime = true;
        options.needsExecution = true;
        options.needsReduceOnly = true;
        return options;
    }
    getSLTPLimitTypeFormOptions(formInputs) {
        const options = {};
        options.needsPrice = true;
        options.needsTriggerPrice = true;
        options.needsGoodTilTime = true;
        options.needsExecution = true;
        options.needsReduceOnly = formInputs.execution === shared_1.OrderExecution.IOC;
        options.executionOptions = [
            shared_1.OrderExecution.DEFAULT,
            shared_1.OrderExecution.IOC,
            shared_1.OrderExecution.POST_ONLY,
        ];
        return options;
    }
    getNeedsMarginMode(formInputs) {
        const marketId = formInputs.marketId;
        const store = this.store.getState();
        const { subaccountNumber, account } = store.user;
        const subaccount = account.subaccounts[subaccountNumber];
        const markets = store.markets;
        const existingMarginMode = this.getExistingMarginMode(marketId, subaccount);
        if (existingMarginMode) {
            return false;
        }
        if (marketId) {
            const market = markets[marketId];
            const marketType = market === null || market === void 0 ? void 0 : market.configs.perpetualMarketType;
            if (marketType === types_1.PerpetualMarketType.CROSS) {
                // if the market is cross, then the user can select either margin mode
                return true;
            }
            // if the market is isolated, then the user can't select cross margin mode
            return false;
        }
        return true;
    }
    getExistingMarginMode(marketId, subaccount) {
        var _a, _b, _c;
        const position = (_a = subaccount === null || subaccount === void 0 ? void 0 : subaccount.openPositions) === null || _a === void 0 ? void 0 : _a[marketId];
        if (position) {
            return (_b = position.marginMode) !== null && _b !== void 0 ? _b : null;
        }
        const openOrder = (_c = subaccount === null || subaccount === void 0 ? void 0 : subaccount.orders) === null || _c === void 0 ? void 0 : _c.find((order) => order.marketId === marketId &&
            order.status.value === shared_1.OrderStatusEnum.OPEN);
        if (openOrder) {
            if (openOrder.subaccountNumber !== (subaccount === null || subaccount === void 0 ? void 0 : subaccount.subaccountNumber)) {
                // if the order is not from the subaccount, then it's isolated
                return types_1.MarginMode.ISOLATED;
            }
            // if the order is from the subaccount, then it's cross
            return types_1.MarginMode.CROSS;
        }
        return null;
    }
}
exports.TradeFormOptionsCalculator = TradeFormOptionsCalculator;
