// @klyra/cortex/dist/src/processors/forms/trade/values.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DEFAULT_TRADE_FORM_STATE = exports.DEFAULT_TRADE_FORM_OPTIONS = void 0;
const shared_1 = require("@klyra/shared");
const forms_1 = require("../../../types/forms");
const accounts_1 = require("../../../types/accounts");
exports.DEFAULT_TRADE_FORM_OPTIONS = {
    needsPrice: false,
    needsTriggerPrice: false,
    needsPostOnly: false,
    needsReduceOnly: false,
    needsTimeInForce: false,
    timeInForceOptions: [shared_1.OrderTimeInForce.IOC, shared_1.OrderTimeInForce.GTT],
    needsGoodTilTime: false,
    needsExecution: false,
    executionOptions: [shared_1.OrderExecution.IOC],
    needsMarginMode: false,
};
exports.DEFAULT_TRADE_FORM_STATE = {
    inputs: {
        type: forms_1.TradeFormType.LIMIT,
        marketId: "",
        side: shared_1.OrderSide.BUY,
        baseSize: 0,
        quoteSize: 0,
        marginMode: accounts_1.MarginMode.CROSS,
        timeInForce: shared_1.OrderTimeInForce.GTT,
        price: 0,
    },
    postResults: {},
    errors: {},
    summary: null,
    options: {
        // since the form is initialized with a GTT time in force and LIMIT type,
        // we need to set the options accordingly
        ...exports.DEFAULT_TRADE_FORM_OPTIONS,
        needsPrice: true,
        needsTimeInForce: true,
        needsGoodTilTime: true,
        needsPostOnly: true,
        needsReduceOnly: false,
    },
};
