// @klyra/proto/src/codegen/klyraprotocol/ratelimit/query.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetAssetYieldIndexQueryResponse = exports.GetAssetYieldIndexQueryRequest = exports.GetSDAIPriceQueryResponse = exports.GetSDAIPriceQueryRequest = exports.QueryAllPendingSendPacketsResponse = exports.QueryAllPendingSendPacketsRequest = exports.QueryCapacityByDenomResponse = exports.QueryCapacityByDenomRequest = exports.ListLimitParamsResponse = exports.ListLimitParamsRequest = void 0;
const limit_params_1 = require("./limit_params");
const capacity_1 = require("./capacity");
const pending_send_packet_1 = require("./pending_send_packet");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseListLimitParamsRequest() {
    return {};
}
exports.ListLimitParamsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListLimitParamsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseListLimitParamsRequest();
        return message;
    }
};
function createBaseListLimitParamsResponse() {
    return {
        limitParamsList: []
    };
}
exports.ListLimitParamsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.limitParamsList) {
            limit_params_1.LimitParams.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseListLimitParamsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.limitParamsList.push(limit_params_1.LimitParams.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseListLimitParamsResponse();
        message.limitParamsList = ((_a = object.limitParamsList) === null || _a === void 0 ? void 0 : _a.map(e => limit_params_1.LimitParams.fromPartial(e))) || [];
        return message;
    }
};
function createBaseQueryCapacityByDenomRequest() {
    return {
        denom: ""
    };
}
exports.QueryCapacityByDenomRequest = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.denom !== "") {
            writer.uint32(10).string(message.denom);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryCapacityByDenomRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.denom = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryCapacityByDenomRequest();
        message.denom = (_a = object.denom) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
function createBaseQueryCapacityByDenomResponse() {
    return {
        limiterCapacityList: []
    };
}
exports.QueryCapacityByDenomResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.limiterCapacityList) {
            capacity_1.LimiterCapacity.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryCapacityByDenomResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.limiterCapacityList.push(capacity_1.LimiterCapacity.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryCapacityByDenomResponse();
        message.limiterCapacityList = ((_a = object.limiterCapacityList) === null || _a === void 0 ? void 0 : _a.map(e => capacity_1.LimiterCapacity.fromPartial(e))) || [];
        return message;
    }
};
function createBaseQueryAllPendingSendPacketsRequest() {
    return {};
}
exports.QueryAllPendingSendPacketsRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAllPendingSendPacketsRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseQueryAllPendingSendPacketsRequest();
        return message;
    }
};
function createBaseQueryAllPendingSendPacketsResponse() {
    return {
        pendingSendPackets: []
    };
}
exports.QueryAllPendingSendPacketsResponse = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.pendingSendPackets) {
            pending_send_packet_1.PendingSendPacket.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseQueryAllPendingSendPacketsResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.pendingSendPackets.push(pending_send_packet_1.PendingSendPacket.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseQueryAllPendingSendPacketsResponse();
        message.pendingSendPackets = ((_a = object.pendingSendPackets) === null || _a === void 0 ? void 0 : _a.map(e => pending_send_packet_1.PendingSendPacket.fromPartial(e))) || [];
        return message;
    }
};
function createBaseGetSDAIPriceQueryRequest() {
    return {};
}
exports.GetSDAIPriceQueryRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSDAIPriceQueryRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseGetSDAIPriceQueryRequest();
        return message;
    }
};
function createBaseGetSDAIPriceQueryResponse() {
    return {
        price: ""
    };
}
exports.GetSDAIPriceQueryResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.price !== "") {
            writer.uint32(10).string(message.price);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetSDAIPriceQueryResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.price = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetSDAIPriceQueryResponse();
        message.price = (_a = object.price) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
function createBaseGetAssetYieldIndexQueryRequest() {
    return {};
}
exports.GetAssetYieldIndexQueryRequest = {
    encode(_, writer = _m0.Writer.create()) {
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAssetYieldIndexQueryRequest();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(_) {
        const message = createBaseGetAssetYieldIndexQueryRequest();
        return message;
    }
};
function createBaseGetAssetYieldIndexQueryResponse() {
    return {
        assetYieldIndex: ""
    };
}
exports.GetAssetYieldIndexQueryResponse = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.assetYieldIndex !== "") {
            writer.uint32(10).string(message.assetYieldIndex);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseGetAssetYieldIndexQueryResponse();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.assetYieldIndex = reader.string();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseGetAssetYieldIndexQueryResponse();
        message.assetYieldIndex = (_a = object.assetYieldIndex) !== null && _a !== void 0 ? _a : "";
        return message;
    }
};
