// @klyra/chain/dist/src/indexer/account.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rest_client_1 = __importDefault(require("../lib/rest-client"));
// TODO: add comments to the methods that explain what they do
class AccountClient extends rest_client_1.default {
    async getSubaccounts(address, limit) {
        return this.get(`/v4/addresses/${address}`, { limit });
    }
    async getSubaccount(address, subaccountNumber) {
        return this.get(`/v4/addresses/${address}/subaccountNumber/${subaccountNumber}`);
    }
    async getSubaccountPerpetualPositions(address, subaccountNumber, status, limit, createdBeforeOrAtHeight, createdBeforeOrAt) {
        return this.get("/v4/perpetualPositions", {
            address,
            subaccountNumber,
            status,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
        });
    }
    async getSubaccountAssetPositions(address, subaccountNumber, status, limit, createdBeforeOrAtHeight, createdBeforeOrAt) {
        return this.get("/v4/assetPositions", {
            address,
            subaccountNumber,
            status,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
        });
    }
    async getSubaccountTransfers(address, subaccountNumber, limit, createdBeforeOrAtHeight, createdBeforeOrAt, page) {
        return this.get("/v4/transfers", {
            address,
            subaccountNumber,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
            page,
        });
    }
    async getParentSubaccountNumberTransfers(address, parentSubaccountNumber, limit, createdBeforeOrAtHeight, createdBeforeOrAt, page) {
        return this.get("/v4/transfers/parentSubaccountNumber", {
            address,
            parentSubaccountNumber,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
            page,
        });
    }
    async getSubaccountOrders(address, subaccountNumber, ticker, side, status, type, limit, goodTilBlockBeforeOrAt, goodTilBlockTimeBeforeOrAt, returnLatestOrders) {
        return this.get("/v4/orders", {
            address,
            subaccountNumber,
            ticker,
            side,
            status,
            type,
            limit,
            goodTilBlockBeforeOrAt,
            goodTilBlockTimeBeforeOrAt,
            returnLatestOrders,
        });
    }
    async getOrder(orderId) {
        return this.get(`/v4/orders/${orderId}`);
    }
    async getSubaccountFills(address, subaccountNumber, ticker, limit, createdBeforeOrAtHeight, createdBeforeOrAt, page) {
        return this.get("/v4/fills", {
            address,
            subaccountNumber,
            ticker,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
            page,
        });
    }
    async getParentSubaccountNumberFills(address, parentSubaccountNumber, ticker, limit, createdBeforeOrAtHeight, createdBeforeOrAt, page) {
        return this.get("/v4/fills/parentSubaccountNumber", {
            address,
            parentSubaccountNumber,
            ticker,
            limit,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
            page,
        });
    }
    async getSubaccountHistoricalPNLs(address, subaccountNumber, createdBeforeOrAtHeight, createdBeforeOrAt, createdOnOrAfterHeight, createdOnOrAfter, limit, page) {
        return this.get("/v4/historical-pnl", {
            address,
            subaccountNumber,
            createdBeforeOrAtHeight,
            createdBeforeOrAt,
            createdOnOrAfterHeight,
            createdOnOrAfter,
            limit,
            page,
        });
    }
}
exports.default = AccountClient;
