// @klyra/chain/dist/src/indexer/markets.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const rest_client_1 = __importDefault(require("../lib/rest-client"));
// TODO: add comments to the methods that explain what they do
// TODO(UCC) - add error handleing if for example you call getPerpetualMarket and the market doesn't exist
class MarketsClient extends rest_client_1.default {
    /**
     * @description Get all perpetual markets
     *
     * @returns All perpetual markets
     */
    async getAllPerpetualMarkets() {
        return this.get("/v4/perpetualMarkets");
    }
    /**
     * @description Get a perpetual market by its id
     *
     * @param marketId The id of the market to get
     * @returns The perpetual market
     */
    async getPerpetualMarket(marketId) {
        const markets = await this.get(`/v4/perpetualMarkets`, {
            ticker: marketId,
        });
        if (!markets.markets[marketId]) {
            throw new Error(`Market ${marketId} not found`);
        }
        return markets.markets[marketId];
    }
    /**
     * @description Get the orderbook for a market
     *
     * @param marketId The id of the market to get the orderbook for
     * @returns The orderbook for the market
     */
    async getPerpetualMarketOrderbook(marketId) {
        return this.get(`/v4/orderbooks/perpetualMarket/${marketId}`);
    }
    /**
     * @description Get the trades for a market
     *
     * @param marketId The id of the market to get the trades for
     * @returns The trades for the market
     */
    async getPerpetualMarketTrades(market, startingBeforeOrAtHeight, startingBeforeOrAt, limit, page) {
        const uri = `/v4/trades/perpetualMarket/${market}`;
        return this.get(uri, {
            createdBeforeOrAtHeight: startingBeforeOrAtHeight,
            createdBeforeOrAt: startingBeforeOrAt,
            limit,
            page,
        });
    }
    /**
     * @description Get the candles for a market
     *
     * @param marketId The id of the market to get the candles for
     * @returns The candles for the market
     */
    async getPerpetualMarketCandles(market, resolution, fromISO, toISO, limit) {
        const uri = `/v4/candles/perpetualMarkets/${market}`;
        return this.get(uri, {
            resolution,
            fromISO,
            toISO,
            limit,
        });
    }
    /**
     * @description Get the historical funding for a perpetual market
     *
     * @param market The id of the market to get the historical funding for
     * @returns The historical funding for the perpetual market
     */
    async getPerpetualMarketHistoricalFunding(market, effectiveBeforeOrAt, effectiveBeforeOrAtHeight, limit) {
        const uri = `/v4/historicalFunding/${market}`;
        return this.get(uri, {
            effectiveBeforeOrAt,
            effectiveBeforeOrAtHeight,
            limit,
        });
    }
    /**
     * @description Get the sparklines for a market.
     * This is a list of the last 24 hours of candle data
     *
     * @param period The period to get the sparklines for
     * @returns The sparklines for the market
     */
    async getPerpetualMarketSparklines(period) {
        const uri = "/v4/sparklines";
        return this.get(uri, {
            timePeriod: period,
        });
    }
}
exports.default = MarketsClient;
