// @klyra/chain/dist/src/node/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const tendermint_rpc_1 = require("@cosmjs/tendermint-rpc");
const stargate_1 = require("@cosmjs/stargate");
const tendermint_1 = require("../lib/tendermint");
const constants_1 = require("../constants");
const node_get_1 = __importDefault(require("./node-get"));
const node_post_1 = __importDefault(require("./node-post"));
class NodeClient {
    constructor(config) {
        this.config = config;
    }
    async initialize() {
        var _a, _b, _c, _d;
        const tendermint37Client = await tendermint_rpc_1.Tendermint37Client.connect(this.config.restUrl);
        const broadcastOptions = {
            pollIntervalMs: (_b = (_a = this.config.broadcastOptions) === null || _a === void 0 ? void 0 : _a.pollIntervalMs) !== null && _b !== void 0 ? _b : constants_1.DEFAULT_BROADCAST_OPTIONS.pollIntervalMs,
            timeoutMs: (_d = (_c = this.config.broadcastOptions) === null || _c === void 0 ? void 0 : _c.timeoutMs) !== null && _d !== void 0 ? _d : constants_1.DEFAULT_BROADCAST_OPTIONS.timeoutMs,
        };
        const tendermintClient = new tendermint_1.TendermintClient(tendermint37Client, broadcastOptions);
        const queryClient = stargate_1.QueryClient.withExtensions(tendermint37Client, stargate_1.setupTxExtension);
        this._get = new node_get_1.default(tendermintClient, queryClient);
        this._post = new node_post_1.default(this._get, this.config.chainId, this.config.defaultClientMemo, this.config.useTimestampNonce);
    }
    get get() {
        if (!this._get) {
            throw new Error("Get module not initialized in NodeClient.");
        }
        return this._get;
    }
    get post() {
        if (!this._post) {
            throw new Error("Post module not initialized in NodeClient.");
        }
        return this._post;
    }
    /**
     * @description populate account number cache in the Post module for performance.
     */
    async populateAccountNumberCache(address) {
        if (!this._post) {
            throw new Error("Post module not initialized in NodeClient.");
        }
        await this._post.populateAccountNumberCache(address);
    }
}
exports.default = NodeClient;
