// @klyra/cortex/dist/src/core/state-manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.StateManager = void 0;
const message_processor_1 = require("../processors/message.processor");
const form_processor_factory_1 = require("../processors/forms/form-processor.factory");
const store_1 = require("./store");
class StateManager {
    constructor(config) {
        this.handleMessage = (message) => {
            this.messageProcessor.process(message);
        };
        this.store = new store_1.Store({
            currentMarketTicker: config.defaultMarketTicker,
        });
        this.messageProcessor = new message_processor_1.MessageProcessor(this.store);
        this.formProcessorFactory = new form_processor_factory_1.FormProcessorFactory(this.store);
    }
    setCurrentMarketTicker(marketId) {
        this.store.setState({
            currentMarketTicker: marketId,
        });
    }
    subscribeToState(listener) {
        return this.store.subscribe(listener);
    }
    getState() {
        return this.store.getState();
    }
    formInput(formType, fieldName, value) {
        const formProcessor = this.formProcessorFactory.getProcessor(formType);
        formProcessor.processInput(fieldName, value);
    }
    setAddress(address) {
        const current = this.getState().user;
        this.store.setState({
            user: {
                ...current,
                address,
            },
        });
    }
    setSubaccountNumber(subaccountNumber) {
        const current = this.getState().user;
        this.store.setState({
            user: { ...current, subaccountNumber },
        });
    }
    setWalletInfo(localWallet) {
        const current = this.getState().user;
        this.store.setState({
            user: {
                ...current,
                wallet: localWallet,
            },
        });
    }
}
exports.StateManager = StateManager;
