// @klyra/cortex/dist/src/core/store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Store = void 0;
const shared_1 = require("@klyra/shared");
const values_1 = require("../processors/forms/trade/values");
const forms_1 = require("../types/forms");
// TODO: add chain info (block height)
const DEFAULT_STATE = {
    forms: {
        [forms_1.FormType.TRADE]: values_1.DEFAULT_TRADE_FORM_STATE,
        [forms_1.FormType.DEPOSIT]: {
            inputs: {
                amount: 0,
                assetId: "",
            },
            postResults: {},
            errors: {},
            summary: null,
            options: null,
        },
        [forms_1.FormType.WITHDRAW]: {
            inputs: {
                amount: 0,
                assetId: "",
            },
            postResults: {},
            errors: {},
            summary: null,
            options: null,
        },
        [forms_1.FormType.TRANSFER]: {
            inputs: {
                amount: 0,
                assetId: "",
            },
            postResults: {},
            errors: {},
            summary: null,
            options: null,
        },
        [forms_1.FormType.CLOSE_POSITION]: {
            inputs: {
                amount: 0,
                assetId: "",
            },
            postResults: {},
            errors: {},
            summary: null,
            options: null,
        },
    },
    user: {
        wallet: null,
        account: {
            subaccounts: {},
            balances: {},
        },
        feeTier: {
            feeTierId: "",
            makerFeeRate: 0,
            takerFeeRate: 0,
            makerVolume30D: 0,
            takerVolume30D: 0,
        },
        address: null,
        subaccountNumber: 0,
    },
    markets: {},
    candles: {},
    orderbooks: {},
    historicalFundings: {},
    trades: {},
    chartConfigs: {
        resolution: "",
    },
    currentMarketTicker: undefined,
};
class Store {
    constructor(initialState = {}) {
        this.state = { ...DEFAULT_STATE, ...initialState };
        this.eventEmitter = new shared_1.EventEmitter();
    }
    getState() {
        return this.state;
    }
    setState(newState) {
        const oldState = this.state;
        this.state = { ...this.state, ...newState };
        this.eventEmitter.emit("stateChange", this.state, oldState);
    }
    subscribe(listener) {
        this.eventEmitter.on("stateChange", listener);
        return () => {
            this.eventEmitter.off("stateChange", listener);
        };
    }
    select(selector) {
        return selector(this.state);
    }
    resetState(newState = {}) {
        this.state = { ...DEFAULT_STATE, ...newState };
        this.eventEmitter.emit("stateChange", this.state, null);
    }
}
exports.Store = Store;
