// @klyra/cortex/src/processors/forms/trade/values.ts
import { OrderSide, OrderTimeInForce, OrderExecution } from "@klyra/shared";
import {
  type BaseForm,
  TradeFormType,
  type TradeFormInputs,
  type TradeFormSummary,
  type TradeFormOptions,
} from "../../../types/forms";
import { MarginMode } from "../../../types/accounts";

export const DEFAULT_TRADE_FORM_OPTIONS: TradeFormOptions = {
  needsPrice: false,
  needsTriggerPrice: false,
  needsPostOnly: false,
  needsReduceOnly: false,
  needsTimeInForce: false,
  timeInForceOptions: [OrderTimeInForce.IOC, OrderTimeInForce.GTT],
  needsGoodTilTime: false,
  needsExecution: false,
  executionOptions: [OrderExecution.IOC],
  needsMarginMode: false,
};

export const DEFAULT_TRADE_FORM_STATE: BaseForm<
  TradeFormInputs,
  TradeFormSummary,
  TradeFormOptions
> = {
  inputs: {
    type: TradeFormType.LIMIT,
    marketId: "",
    side: OrderSide.BUY,
    baseSize: 0,
    quoteSize: 0,
    marginMode: MarginMode.CROSS,
    timeInForce: OrderTimeInForce.GTT,
    price: 0,
  },
  postResults: {},
  errors: {},
  summary: null,
  options: {
    // since the form is initialized with a GTT time in force and LIMIT type,
    // we need to set the options accordingly
    ...DEFAULT_TRADE_FORM_OPTIONS,
    needsPrice: true,
    needsTimeInForce: true,
    needsGoodTilTime: true,
    needsPostOnly: true,
    needsReduceOnly: false,
  },
};
