// @klyra/cortex/dist/src/utils/indexer-conversion.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.convertIndexerOrderType = convertIndexerOrderType;
exports.convertIndexerLiquidity = convertIndexerLiquidity;
exports.convertIndexerOrderStatus = convertIndexerOrderStatus;
exports.convertIndexerTimeInForce = convertIndexerTimeInForce;
exports.convertIndexerOrderSide = convertIndexerOrderSide;
exports.convertIndexerPositionSide = convertIndexerPositionSide;
exports.getMarginModeFromSubaccount = getMarginModeFromSubaccount;
const shared_1 = require("@klyra/shared");
const shared_2 = require("@klyra/shared");
const accounts_1 = require("../types/accounts");
const order_status_1 = require("./order-status");
function convertIndexerOrderType(type) {
    // TODO: check if the order types are correct
    switch (type) {
        case shared_1.IndexerFillType.LIMIT:
            return shared_1.OrderType.LIMIT;
        case shared_1.IndexerFillType.LIQUIDATED:
            return shared_1.OrderType.LIQUIDATED;
        case shared_1.IndexerFillType.LIQUIDATION:
            return shared_1.OrderType.LIQUIDATION;
        default:
            return shared_1.OrderType.LIMIT;
    }
}
function convertIndexerLiquidity(liquidity) {
    if (liquidity === shared_1.IndexerLiquidity.MAKER) {
        return accounts_1.FillLiquidity.MAKER;
    }
    return accounts_1.FillLiquidity.TAKER;
}
function convertIndexerOrderStatus(status) {
    if (status === shared_1.IndexerOrderStatus.BEST_EFFORT_OPENED) {
        return (0, order_status_1.getOrderStatusFromEnum)(shared_1.OrderStatusEnum.PENDING);
    }
    const value = status;
    return (0, order_status_1.getOrderStatusFromEnum)(value);
}
function convertIndexerTimeInForce(timeInForce) {
    if (timeInForce === shared_1.IndexerTimeInForce.GTT) {
        return shared_1.OrderTimeInForce.GTT;
    }
    if (timeInForce === shared_1.IndexerTimeInForce.IOC) {
        return shared_1.OrderTimeInForce.IOC;
    }
    return null;
}
function convertIndexerOrderSide(side) {
    return side === shared_1.IndexerOrderSide.BUY ? shared_1.OrderSide.BUY : shared_1.OrderSide.SELL;
}
function convertIndexerPositionSide(side) {
    return side === shared_1.IndexerPositionSide.LONG
        ? accounts_1.PositionSide.LONG
        : accounts_1.PositionSide.SHORT;
}
function getMarginModeFromSubaccount(subaccountNumber) {
    return subaccountNumber >= shared_2.NUM_PARENT_SUBACCOUNTS
        ? accounts_1.MarginMode.ISOLATED
        : accounts_1.MarginMode.CROSS;
}
