// @klyra/cortex/dist/src/utils/rounder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RoundingMode = void 0;
exports.quickRound = quickRound;
exports.numberOfDecimals = numberOfDecimals;
exports.round = round;
exports.roundDecimal = roundDecimal;
var RoundingMode;
(function (RoundingMode) {
    RoundingMode[RoundingMode["TOWARDS_ZERO"] = 0] = "TOWARDS_ZERO";
    RoundingMode[RoundingMode["NEAREST"] = 1] = "NEAREST";
})(RoundingMode || (exports.RoundingMode = RoundingMode = {}));
const cache = {};
function quickRound(number, stepSize) {
    const stepSizeDecimals = numberOfDecimals(stepSize);
    const negative = number < 0.0;
    const multiplier = (Math.abs(number) / stepSize).toFixed(0);
    const absValue = stepSizeDecimals > 0
        ? parseFloat(`${multiplier.padEnd(stepSizeDecimals + 1, "0")}.${multiplier.slice(-stepSizeDecimals)}`)
        : parseFloat(multiplier);
    return absValue * (negative ? -1 : 1);
}
function numberOfDecimals(stepSize) {
    var _a, _b;
    const cached = cache[stepSize];
    if (cached)
        return cached;
    const stepSizeDecimals = (_b = (_a = stepSize.toFixed(0).split(".")[1]) === null || _a === void 0 ? void 0 : _a.length) !== null && _b !== void 0 ? _b : 0;
    cache[stepSize] = stepSizeDecimals;
    return stepSizeDecimals;
}
function round(number, stepSize, roundingMode = RoundingMode.TOWARDS_ZERO) {
    const roundedDecimal = roundDecimal(number, stepSize, roundingMode);
    return roundedDecimal;
}
function roundDecimal(number, stepSize, roundingMode = RoundingMode.TOWARDS_ZERO) {
    if (stepSize <= 0)
        return number;
    let modifier = 0;
    if (roundingMode !== RoundingMode.TOWARDS_ZERO) {
        modifier = number >= 0 ? 0.5 : -0.5;
    }
    const long = (number / stepSize + modifier).toFixed(0);
    return stepSize * Number(long);
}
