// @klyra/proto/src/codegen/klyraprotocol/clob/liquidations_config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.FillablePriceConfig = exports.LiquidationsConfig = void 0;
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseLiquidationsConfig() {
    return {
        insuranceFundFeePpm: 0,
        validatorFeePpm: 0,
        liquidityFeePpm: 0,
        fillablePriceConfig: undefined,
        maxCumulativeInsuranceFundDelta: helpers_1.Long.UZERO
    };
}
exports.LiquidationsConfig = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.insuranceFundFeePpm !== 0) {
            writer.uint32(8).uint32(message.insuranceFundFeePpm);
        }
        if (message.validatorFeePpm !== 0) {
            writer.uint32(16).uint32(message.validatorFeePpm);
        }
        if (message.liquidityFeePpm !== 0) {
            writer.uint32(24).uint32(message.liquidityFeePpm);
        }
        if (message.fillablePriceConfig !== undefined) {
            exports.FillablePriceConfig.encode(message.fillablePriceConfig, writer.uint32(34).fork()).ldelim();
        }
        if (!message.maxCumulativeInsuranceFundDelta.isZero()) {
            writer.uint32(40).uint64(message.maxCumulativeInsuranceFundDelta);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseLiquidationsConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.insuranceFundFeePpm = reader.uint32();
                    break;
                case 2:
                    message.validatorFeePpm = reader.uint32();
                    break;
                case 3:
                    message.liquidityFeePpm = reader.uint32();
                    break;
                case 4:
                    message.fillablePriceConfig = exports.FillablePriceConfig.decode(reader, reader.uint32());
                    break;
                case 5:
                    message.maxCumulativeInsuranceFundDelta = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b, _c;
        const message = createBaseLiquidationsConfig();
        message.insuranceFundFeePpm = (_a = object.insuranceFundFeePpm) !== null && _a !== void 0 ? _a : 0;
        message.validatorFeePpm = (_b = object.validatorFeePpm) !== null && _b !== void 0 ? _b : 0;
        message.liquidityFeePpm = (_c = object.liquidityFeePpm) !== null && _c !== void 0 ? _c : 0;
        message.fillablePriceConfig = object.fillablePriceConfig !== undefined && object.fillablePriceConfig !== null ? exports.FillablePriceConfig.fromPartial(object.fillablePriceConfig) : undefined;
        message.maxCumulativeInsuranceFundDelta = object.maxCumulativeInsuranceFundDelta !== undefined && object.maxCumulativeInsuranceFundDelta !== null ? helpers_1.Long.fromValue(object.maxCumulativeInsuranceFundDelta) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseFillablePriceConfig() {
    return {
        bankruptcyAdjustmentPpm: 0,
        spreadToMaintenanceMarginRatioPpm: 0
    };
}
exports.FillablePriceConfig = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.bankruptcyAdjustmentPpm !== 0) {
            writer.uint32(8).uint32(message.bankruptcyAdjustmentPpm);
        }
        if (message.spreadToMaintenanceMarginRatioPpm !== 0) {
            writer.uint32(16).uint32(message.spreadToMaintenanceMarginRatioPpm);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseFillablePriceConfig();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.bankruptcyAdjustmentPpm = reader.uint32();
                    break;
                case 2:
                    message.spreadToMaintenanceMarginRatioPpm = reader.uint32();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseFillablePriceConfig();
        message.bankruptcyAdjustmentPpm = (_a = object.bankruptcyAdjustmentPpm) !== null && _a !== void 0 ? _a : 0;
        message.spreadToMaintenanceMarginRatioPpm = (_b = object.spreadToMaintenanceMarginRatioPpm) !== null && _b !== void 0 ? _b : 0;
        return message;
    }
};
