// @klyra/shared/dist/utils/event-emitter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventEmitter = void 0;
class EventEmitter {
    constructor() {
        this.events = {};
    }
    on(event, listener) {
        var _a;
        var _b;
        ((_a = (_b = this.events)[event]) !== null && _a !== void 0 ? _a : (_b[event] = [])).push(listener);
    }
    off(event, listenerToRemove) {
        var _a;
        const listeners = (_a = this.events[event]) !== null && _a !== void 0 ? _a : [];
        this.events[event] = listeners.filter((listener) => listener !== listenerToRemove);
    }
    emit(event, ...args) {
        var _a;
        const listeners = (_a = this.events[event]) !== null && _a !== void 0 ? _a : [];
        listeners.forEach((listener) => {
            listener.apply(this, args);
        });
    }
    once(event, listener) {
        const onceWrapper = (...args) => {
            listener.apply(this, args);
            this.off(event, onceWrapper);
        };
        this.on(event, onceWrapper);
    }
    removeAllListeners(event) {
        if (event) {
            this.events[event] = [];
        }
        else {
            this.events = {};
        }
    }
    listenerCount(event) {
        var _a;
        const listeners = (_a = this.events[event]) !== null && _a !== void 0 ? _a : [];
        return listeners.length;
    }
    rawListeners(event) {
        var _a;
        return (_a = this.events[event]) !== null && _a !== void 0 ? _a : [];
    }
}
exports.EventEmitter = EventEmitter;
