// @klyra/cortex/dist/src/processors/candles.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CandlesProcessor = void 0;
const merge_1 = require("../utils/merge");
class CandlesProcessor {
    constructor(store) {
        this.store = store;
    }
    processSubscribed(msg) {
        const { marketId, resolution } = this.getMarketAndResolution(msg.id);
        const payload = msg.contents.candles;
        const modified = payload.reverse().map(this.processCandle.bind(this));
        this.updateCandles(marketId, resolution, modified);
    }
    // Even though this is a batch update, we only want to update the last candle
    // because the indexer will send the latest candle for a given resolution
    // and time period.
    processBatchUpdate(msg) {
        var _a, _b, _c, _d;
        const { marketId, resolution } = this.getMarketAndResolution(msg.id);
        const existingCandles = this.store.getState().candles;
        const payload = msg.contents;
        const newCandlesArray = (_b = (_a = existingCandles[marketId]) === null || _a === void 0 ? void 0 : _a[resolution]) !== null && _b !== void 0 ? _b : [];
        const lastExistingCandle = (_c = newCandlesArray[newCandlesArray.length - 1]) !== null && _c !== void 0 ? _c : null;
        const lastExistingCandleTimestamp = (_d = lastExistingCandle === null || lastExistingCandle === void 0 ? void 0 : lastExistingCandle.startedAtMilliseconds) !== null && _d !== void 0 ? _d : 0;
        for (const candle of payload) {
            const newCandle = this.processCandle(candle);
            const newCandleTimestamp = newCandle.startedAtMilliseconds;
            if (newCandleTimestamp === lastExistingCandleTimestamp) {
                newCandlesArray.pop();
            }
            newCandlesArray.push(newCandle);
        }
        this.updateCandles(marketId, resolution, newCandlesArray);
    }
    getMarketAndResolution(id) {
        if (!id)
            throw new Error("No id provided");
        const [marketId, resolution] = id.split("/");
        const resolutionEnum = resolution;
        return { marketId: marketId, resolution: resolutionEnum };
    }
    processCandle(payload) {
        return (0, merge_1.transform)(undefined, payload, {
            number: {
                low: "low",
                high: "high",
                open: "open",
                close: "close",
                baseTokenVolume: "baseTokenVolume",
                usdVolume: "usdVolume",
                trades: "trades",
            },
            datetime: {
                startedAt: "startedAtMilliseconds",
                updatedAt: "updatedAtMilliseconds",
            },
        });
    }
    updateCandles(marketId, resolution, candles) {
        const existingCandles = this.store.getState().candles;
        this.store.setState({
            candles: {
                ...existingCandles,
                [marketId]: {
                    ...existingCandles[marketId],
                    [resolution]: candles,
                },
            },
        });
    }
}
exports.CandlesProcessor = CandlesProcessor;
