// @klyra/cortex/dist/src/processors/message.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageProcessor = void 0;
const shared_1 = require("@klyra/shared");
const markets_processor_1 = require("./markets.processor");
const trades_processor_1 = require("./trades.processor");
const orderbooks_processor_1 = require("./orderbooks.processor");
const candles_processor_1 = require("./candles.processor");
const account_processor_1 = require("./account/account.processor");
class MessageProcessor {
    constructor(store) {
        this.accountProcessor = new account_processor_1.AccountProcessor(store);
        this.marketsProcessor = new markets_processor_1.MarketsProcessor(store);
        this.tradesProcessor = new trades_processor_1.TradesProcessor(store);
        this.orderbooksProcessor = new orderbooks_processor_1.OrderbooksProcessor(store);
        this.candlesProcessor = new candles_processor_1.CandlesProcessor(store);
    }
    process(message) {
        if (message.channel.length === 0)
            return;
        switch (message.channel) {
            case shared_1.MessageChannel.MARKETS:
                this.processMarketUpdate(message);
                break;
            case shared_1.MessageChannel.TRADES:
                this.processTradeUpdate(message);
                break;
            case shared_1.MessageChannel.ORDERBOOK:
                this.processOrderbookUpdate(message);
                break;
            case shared_1.MessageChannel.CANDLES:
                this.processCandleUpdate(message);
                break;
            case shared_1.MessageChannel.PARENT_SUBACCOUNTS:
                this.processAccountUpdate(message);
                break;
            default:
                throw new Error(`Unsupported message channel: ${message.channel}`);
        }
    }
    processMarketUpdate(msg) {
        if (msg.type === shared_1.MessageType.CHANNEL_BATCH_DATA) {
            this.marketsProcessor.processBatchUpdate(msg);
        }
        else if (msg.type === shared_1.MessageType.SUBSCRIBED) {
            this.marketsProcessor.processSubscribed(msg);
        }
    }
    processTradeUpdate(msg) {
        if (msg.type === shared_1.MessageType.CHANNEL_BATCH_DATA) {
            this.tradesProcessor.processBatchUpdate(msg);
        }
        else if (msg.type === shared_1.MessageType.SUBSCRIBED) {
            this.tradesProcessor.processSubscribed(msg);
        }
        return {};
    }
    processOrderbookUpdate(msg) {
        if (msg.type === shared_1.MessageType.CHANNEL_BATCH_DATA) {
            this.orderbooksProcessor.processBatchUpdate(msg);
        }
        else if (msg.type === shared_1.MessageType.SUBSCRIBED) {
            this.orderbooksProcessor.processSubscribed(msg);
        }
        return {};
    }
    processAccountUpdate(msg) {
        if (msg.type === shared_1.MessageType.CHANNEL_BATCH_DATA) {
            this.accountProcessor.processBatchUpdate(msg);
        }
        else if (msg.type === shared_1.MessageType.SUBSCRIBED) {
            this.accountProcessor.processSubscribed(msg);
        }
        return {};
    }
    processCandleUpdate(msg) {
        if (msg.type === shared_1.MessageType.CHANNEL_BATCH_DATA) {
            this.candlesProcessor.processBatchUpdate(msg);
        }
        else if (msg.type === shared_1.MessageType.SUBSCRIBED) {
            this.candlesProcessor.processSubscribed(msg);
        }
        return {};
    }
}
exports.MessageProcessor = MessageProcessor;
