// @klyra/cortex/dist/src/processors/trades.processor.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TradesProcessor = void 0;
const merge_1 = require("../utils/merge");
class TradesProcessor {
    constructor(store) {
        // TODO: make this configurable
        this.TRADES_LIMIT = 100;
        this.store = store;
    }
    processSubscribed(msg) {
        const newMarketTrades = [];
        for (const payload of msg.contents.trades) {
            const newTrade = this.processTrade(payload);
            if (newTrade)
                newMarketTrades.push(newTrade);
        }
        const marketId = msg.id;
        this.updateMarketTrades(marketId, newMarketTrades);
    }
    processBatchUpdate(msg) {
        const newMarketTrades = [];
        for (const array of msg.contents) {
            for (const payload of array.trades) {
                const newTrade = this.processTrade(payload);
                if (newTrade)
                    newMarketTrades.push(newTrade);
            }
        }
        const marketId = msg.id;
        this.updateMarketTrades(marketId, newMarketTrades);
    }
    processTrade(payload) {
        const newTrade = (0, merge_1.transform)(undefined, payload, {
            number: {
                price: "price",
                size: "size",
            },
            string: {
                id: "id",
                side: "side",
                type: "type",
            },
            datetime: {
                createdAt: "timestamp",
            },
        });
        return newTrade;
    }
    updateMarketTrades(marketId, newTrades) {
        var _a;
        const existingTrades = this.store.getState().trades;
        const existingMarketTrades = (_a = existingTrades[marketId]) !== null && _a !== void 0 ? _a : [];
        const updatedMarketTrades = [...existingMarketTrades, ...newTrades];
        const limitedMarketTrades = updatedMarketTrades.slice(-this.TRADES_LIMIT);
        this.store.setState({
            trades: {
                ...existingTrades,
                [marketId]: limitedMarketTrades,
            },
        });
    }
}
exports.TradesProcessor = TradesProcessor;
