// @klyra/proto/src/codegen/klyraprotocol/bundle.ts
import * as _135 from "./assets/asset";
import * as _136 from "./assets/genesis";
import * as _137 from "./assets/query";
import * as _138 from "./assets/tx";
import * as _139 from "./blocktime/blocktime";
import * as _140 from "./blocktime/genesis";
import * as _141 from "./blocktime/params";
import * as _142 from "./blocktime/query";
import * as _143 from "./blocktime/tx";
import * as _144 from "./clob/block_rate_limit_config";
import * as _145 from "./clob/clob_pair";
import * as _146 from "./clob/equity_tier_limit_config";
import * as _147 from "./clob/genesis";
import * as _148 from "./clob/liquidations_config";
import * as _149 from "./clob/liquidations";
import * as _150 from "./clob/matches";
import * as _151 from "./clob/mev";
import * as _152 from "./clob/operation";
import * as _153 from "./clob/order_removals";
import * as _154 from "./clob/order";
import * as _155 from "./clob/process_proposer_matches_events";
import * as _156 from "./clob/query";
import * as _157 from "./clob/tx";
import * as _158 from "./daemons/deleveraging/deleveraging";
import * as _159 from "./daemons/pricefeed/price_feed";
import * as _160 from "./daemons/sdaioracle/sdai";
import * as _161 from "./delaymsg/block_message_ids";
import * as _162 from "./delaymsg/delayed_message";
import * as _163 from "./delaymsg/genesis";
import * as _164 from "./delaymsg/query";
import * as _165 from "./delaymsg/tx";
import * as _166 from "./epochs/epoch_info";
import * as _167 from "./epochs/genesis";
import * as _168 from "./epochs/query";
import * as _169 from "./feetiers/genesis";
import * as _170 from "./feetiers/params";
import * as _171 from "./feetiers/query";
import * as _172 from "./feetiers/tx";
import * as _173 from "./govplus/genesis";
import * as _174 from "./govplus/query";
import * as _175 from "./govplus/tx";
import * as _176 from "./indexer/events/events";
import * as _177 from "./indexer/indexer_manager/event";
import * as _178 from "./indexer/off_chain_updates/off_chain_updates";
import * as _179 from "./indexer/protocol/v1/clob";
import * as _180 from "./indexer/protocol/v1/perpetual";
import * as _181 from "./indexer/protocol/v1/subaccount";
import * as _182 from "./indexer/redis/redis_order";
import * as _183 from "./indexer/shared/removal_reason";
import * as _184 from "./indexer/socks/messages";
import * as _185 from "./perpetuals/genesis";
import * as _186 from "./perpetuals/params";
import * as _187 from "./perpetuals/perpetual";
import * as _188 from "./perpetuals/query";
import * as _189 from "./perpetuals/tx";
import * as _190 from "./prices/genesis";
import * as _191 from "./prices/market_param";
import * as _192 from "./prices/market_price";
import * as _193 from "./prices/query";
import * as _194 from "./prices/tx";
import * as _195 from "./ratelimit/capacity";
import * as _196 from "./ratelimit/genesis";
import * as _197 from "./ratelimit/limit_params";
import * as _198 from "./ratelimit/pending_send_packet";
import * as _199 from "./ratelimit/query";
import * as _200 from "./ratelimit/tx";
import * as _201 from "./rewards/genesis";
import * as _202 from "./rewards/params";
import * as _203 from "./rewards/query";
import * as _204 from "./rewards/reward_share";
import * as _205 from "./rewards/tx";
import * as _206 from "./sending/genesis";
import * as _207 from "./sending/query";
import * as _208 from "./sending/transfer";
import * as _209 from "./sending/tx";
import * as _210 from "./stats/genesis";
import * as _211 from "./stats/params";
import * as _212 from "./stats/query";
import * as _213 from "./stats/stats";
import * as _214 from "./stats/tx";
import * as _215 from "./subaccounts/asset_position";
import * as _216 from "./subaccounts/genesis";
import * as _217 from "./subaccounts/perpetual_position";
import * as _218 from "./subaccounts/query";
import * as _219 from "./subaccounts/subaccount";
import * as _220 from "./subaccounts/tx";
import * as _221 from "./ve/ve";
import * as _222 from "./vest/genesis";
import * as _223 from "./vest/query";
import * as _224 from "./vest/tx";
import * as _225 from "./vest/vest_entry";
import * as _297 from "./assets/query.lcd";
import * as _298 from "./blocktime/query.lcd";
import * as _299 from "./clob/query.lcd";
import * as _300 from "./delaymsg/query.lcd";
import * as _301 from "./epochs/query.lcd";
import * as _302 from "./feetiers/query.lcd";
import * as _303 from "./perpetuals/query.lcd";
import * as _304 from "./prices/query.lcd";
import * as _305 from "./ratelimit/query.lcd";
import * as _306 from "./rewards/query.lcd";
import * as _307 from "./stats/query.lcd";
import * as _308 from "./subaccounts/query.lcd";
import * as _309 from "./vest/query.lcd";
import * as _310 from "./assets/query.rpc.Query";
import * as _311 from "./blocktime/query.rpc.Query";
import * as _312 from "./clob/query.rpc.Query";
import * as _313 from "./delaymsg/query.rpc.Query";
import * as _314 from "./epochs/query.rpc.Query";
import * as _315 from "./feetiers/query.rpc.Query";
import * as _316 from "./govplus/query.rpc.Query";
import * as _317 from "./perpetuals/query.rpc.Query";
import * as _318 from "./prices/query.rpc.Query";
import * as _319 from "./ratelimit/query.rpc.Query";
import * as _320 from "./rewards/query.rpc.Query";
import * as _321 from "./sending/query.rpc.Query";
import * as _322 from "./stats/query.rpc.Query";
import * as _323 from "./subaccounts/query.rpc.Query";
import * as _324 from "./vest/query.rpc.Query";
import * as _325 from "./blocktime/tx.rpc.msg";
import * as _326 from "./clob/tx.rpc.msg";
import * as _327 from "./delaymsg/tx.rpc.msg";
import * as _328 from "./feetiers/tx.rpc.msg";
import * as _329 from "./govplus/tx.rpc.msg";
import * as _330 from "./perpetuals/tx.rpc.msg";
import * as _331 from "./prices/tx.rpc.msg";
import * as _332 from "./ratelimit/tx.rpc.msg";
import * as _333 from "./rewards/tx.rpc.msg";
import * as _334 from "./sending/tx.rpc.msg";
import * as _335 from "./stats/tx.rpc.msg";
import * as _336 from "./subaccounts/tx.rpc.msg";
import * as _337 from "./vest/tx.rpc.msg";
import * as _341 from "./lcd";
import * as _342 from "./rpc.query";
import * as _343 from "./rpc.tx";
export namespace klyraprotocol {
  export const assets = { ..._135,
    ..._136,
    ..._137,
    ..._138,
    ..._297,
    ..._310
  };
  export const blocktime = { ..._139,
    ..._140,
    ..._141,
    ..._142,
    ..._143,
    ..._298,
    ..._311,
    ..._325
  };
  export const clob = { ..._144,
    ..._145,
    ..._146,
    ..._147,
    ..._148,
    ..._149,
    ..._150,
    ..._151,
    ..._152,
    ..._153,
    ..._154,
    ..._155,
    ..._156,
    ..._157,
    ..._299,
    ..._312,
    ..._326
  };
  export namespace daemons {
    export const deleveraging = { ..._158
    };
    export const pricefeed = { ..._159
    };
    export const sdaioracle = { ..._160
    };
  }
  export const delaymsg = { ..._161,
    ..._162,
    ..._163,
    ..._164,
    ..._165,
    ..._300,
    ..._313,
    ..._327
  };
  export const epochs = { ..._166,
    ..._167,
    ..._168,
    ..._301,
    ..._314
  };
  export const feetiers = { ..._169,
    ..._170,
    ..._171,
    ..._172,
    ..._302,
    ..._315,
    ..._328
  };
  export const govplus = { ..._173,
    ..._174,
    ..._175,
    ..._316,
    ..._329
  };
  export namespace indexer {
    export const events = { ..._176
    };
    export const indexer_manager = { ..._177
    };
    export const off_chain_updates = { ..._178
    };
    export namespace protocol {
      export const v1 = { ..._179,
        ..._180,
        ..._181
      };
    }
    export const redis = { ..._182
    };
    export const shared = { ..._183
    };
    export const socks = { ..._184
    };
  }
  export const perpetuals = { ..._185,
    ..._186,
    ..._187,
    ..._188,
    ..._189,
    ..._303,
    ..._317,
    ..._330
  };
  export const prices = { ..._190,
    ..._191,
    ..._192,
    ..._193,
    ..._194,
    ..._304,
    ..._318,
    ..._331
  };
  export const ratelimit = { ..._195,
    ..._196,
    ..._197,
    ..._198,
    ..._199,
    ..._200,
    ..._305,
    ..._319,
    ..._332
  };
  export const rewards = { ..._201,
    ..._202,
    ..._203,
    ..._204,
    ..._205,
    ..._306,
    ..._320,
    ..._333
  };
  export const sending = { ..._206,
    ..._207,
    ..._208,
    ..._209,
    ..._321,
    ..._334
  };
  export const stats = { ..._210,
    ..._211,
    ..._212,
    ..._213,
    ..._214,
    ..._307,
    ..._322,
    ..._335
  };
  export const subaccounts = { ..._215,
    ..._216,
    ..._217,
    ..._218,
    ..._219,
    ..._220,
    ..._308,
    ..._323,
    ..._336
  };
  export const ve = { ..._221
  };
  export const vest = { ..._222,
    ..._223,
    ..._224,
    ..._225,
    ..._309,
    ..._324,
    ..._337
  };
  export const ClientFactory = { ..._341,
    ..._342,
    ..._343
  };
}
