// @klyra/chain/dist/src/lib/composer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const long_1 = __importDefault(require("long"));
const protobufjs_1 = __importDefault(require("protobufjs"));
const constants_1 = require("../constants");
const proto_includes_1 = require("./proto-includes");
protobufjs_1.default.util.Long = long_1.default;
protobufjs_1.default.configure();
class Composer {
    // ------------ x/clob ------------
    composeMsgPlaceOrder(address, subaccountNumber, clientId, clobPairId, orderFlags, goodTilBlock, goodTilBlockTime, side, quantums, subticks, routerFeePpm, routerSubaccountAddress, routerSubaccountNumber, timeInForce, reduceOnly, clientMetadata, conditionType = proto_includes_1.Order_ConditionType.CONDITION_TYPE_UNSPECIFIED, conditionalOrderTriggerSubticks = long_1.default.fromInt(0)) {
        this.validateGoodTilBlockAndTime(orderFlags, goodTilBlock, goodTilBlockTime);
        const subaccountId = {
            owner: address,
            number: subaccountNumber,
        };
        const routerSubaccountId = {
            owner: routerSubaccountAddress,
            number: routerSubaccountNumber,
        };
        const orderId = {
            subaccountId,
            clientId,
            orderFlags,
            clobPairId,
        };
        const order = {
            orderId,
            side,
            quantums,
            subticks,
            goodTilBlock: goodTilBlock === 0 ? undefined : goodTilBlock,
            goodTilBlockTime: goodTilBlock === 0 ? goodTilBlockTime : undefined,
            timeInForce,
            reduceOnly,
            clientMetadata,
            conditionType,
            conditionalOrderTriggerSubticks,
            routerFeePpm,
            routerSubaccountId,
        };
        const msg = {
            order,
        };
        return {
            typeUrl: constants_1.Messages.PLACE_ORDER,
            value: msg,
        };
    }
    composeMsgCancelOrder(address, subaccountNumber, clientId, clobPairId, orderFlags, goodTilBlock, goodTilBlockTime) {
        this.validateGoodTilBlockAndTime(orderFlags, goodTilBlock, goodTilBlockTime);
        const subaccountId = {
            owner: address,
            number: subaccountNumber,
        };
        const orderId = {
            subaccountId,
            clientId,
            orderFlags,
            clobPairId,
        };
        const msg = {
            orderId,
            goodTilBlock: goodTilBlock === 0 ? undefined : goodTilBlock,
            goodTilBlockTime: goodTilBlock === 0 ? goodTilBlockTime : undefined,
        };
        return {
            typeUrl: constants_1.Messages.CANCEL_ORDER,
            value: msg,
        };
    }
    composeMsgBatchCancelShortTermOrders(address, subaccountNumber, shortTermCancels, goodTilBlock) {
        const subaccountId = {
            owner: address,
            number: subaccountNumber,
        };
        const msg = {
            subaccountId,
            shortTermCancels,
            goodTilBlock,
        };
        return {
            typeUrl: constants_1.Messages.BATCH_CANCEL,
            value: msg,
        };
    }
    // ------------ x/sending ------------
    composeMsgTransfer(address, subaccountNumber, recipientAddress, recipientSubaccountNumber, assetId, amount) {
        const sender = {
            owner: address,
            number: subaccountNumber,
        };
        const recipient = {
            owner: recipientAddress,
            number: recipientSubaccountNumber,
        };
        const transfer = {
            sender,
            recipient,
            assetId,
            amount,
        };
        const msg = {
            transfer,
        };
        return {
            typeUrl: constants_1.Messages.CREATE_TRANSFER,
            value: msg,
        };
    }
    composeMsgDepositToSubaccount(address, subaccountNumber, assetId, quantums) {
        const recipient = {
            owner: address,
            number: subaccountNumber,
        };
        const msg = {
            sender: address,
            recipient,
            assetId,
            quantums,
        };
        return {
            typeUrl: constants_1.Messages.DEPOSIT_TO_SUBACCOUNT,
            value: msg,
        };
    }
    composeMsgWithdrawFromSubaccount(address, subaccountNumber, assetId, quantums, recipient = address) {
        const sender = {
            owner: address,
            number: subaccountNumber,
        };
        const msg = {
            sender,
            recipient,
            assetId,
            quantums,
        };
        return {
            typeUrl: constants_1.Messages.WITHDRAW_FROM_SUBACCOUNT,
            value: msg,
        };
    }
    // ------------ x/bank ------------
    composeMsgSendToken(address, recipient, coinDenom, quantums) {
        const coin = {
            denom: coinDenom,
            amount: quantums,
        };
        const msg = {
            fromAddress: address,
            toAddress: recipient,
            amount: [coin],
        };
        return {
            typeUrl: constants_1.Messages.SEND,
            value: msg,
        };
    }
    // ------------ util ------------
    validateGoodTilBlockAndTime(orderFlags, goodTilBlock, goodTilBlockTime) {
        if (orderFlags === 0 && goodTilBlock === 0) {
            throw new Error("goodTilBlock must be set if orderFlags is 0");
        }
        else if (orderFlags !== 0 && goodTilBlockTime === 0) {
            throw new Error("goodTilBlockTime must be set if orderFlags is not 0");
        }
    }
    wrapMessageAsAny(registry, message) {
        return registry.encodeAsAny(message);
    }
    wrapMessageArrAsAny(registry, messages) {
        const encodedMessages = messages.map((message) => this.wrapMessageAsAny(registry, message));
        return encodedMessages;
    }
}
exports.default = Composer;
