// @klyra/chain/dist/src/lib/local-wallet.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.LocalWallet = void 0;
const amino_1 = require("@cosmjs/amino");
const proto_signing_1 = require("@cosmjs/proto-signing");
const stargate_1 = require("@cosmjs/stargate");
const long_1 = __importDefault(require("long"));
const protobufjs_1 = __importDefault(require("protobufjs"));
const registry_1 = require("./registry");
const signer_1 = require("./signer");
// Required for encoding and decoding queries that are of type Long.
protobufjs_1.default.util.Long = long_1.default;
protobufjs_1.default.configure();
class LocalWallet {
    constructor(offlineSigner, accounts, address, pubKey, transactionSigner) {
        this.offlineSigner = offlineSigner;
        this.accounts = accounts;
        this.address = address;
        this.pubKey = pubKey;
        this.transactionSigner = transactionSigner;
    }
    static async fromOfflineSigner(signer) {
        const stargateClient = await stargate_1.SigningStargateClient.offline(signer, {
            registry: (0, registry_1.generateRegistry)(),
        });
        const accounts = await signer.getAccounts();
        const firstAccount = accounts[0];
        if (!firstAccount) {
            throw new Error("No accounts found");
        }
        const address = firstAccount.address;
        const pubKey = (0, amino_1.encodeSecp256k1Pubkey)(firstAccount.pubkey);
        const transactionSigner = new signer_1.TransactionSigner(address, stargateClient);
        return new LocalWallet(signer, accounts, address, pubKey, transactionSigner);
    }
    static async fromMnemonic(mnemonic, prefix) {
        const signer = await proto_signing_1.DirectSecp256k1HdWallet.fromMnemonic(mnemonic, {
            prefix,
        });
        return LocalWallet.fromOfflineSigner(signer);
    }
    getAccounts() {
        return this.accounts;
    }
    getAddress() {
        return this.address;
    }
    getPubKey() {
        return this.pubKey;
    }
    async signTransaction(messages, transactionOptions, fee, memo = "") {
        return this.transactionSigner.signTransaction([...messages], transactionOptions, fee, memo);
    }
}
exports.LocalWallet = LocalWallet;
