// @klyra/chain/dist/src/lib/proto-includes.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RateLimitModule = exports.FeeTierModule = exports.SubaccountsModule = exports.PricesModule = exports.PerpetualsModule = exports.ClobModule = exports.StatsModule = void 0;
exports.StatsModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/stats/query"));
exports.ClobModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/clob/query"));
exports.PerpetualsModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/perpetuals/query"));
exports.PricesModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/prices/query"));
exports.SubaccountsModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/subaccounts/query"));
exports.FeeTierModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/feetiers/query"));
exports.RateLimitModule = __importStar(require("@klyra/proto/src/codegen/klyraprotocol/ratelimit/query"));
__exportStar(require("@klyra/proto/src/codegen/cosmos/base/abci/v1beta1/abci"), exports);
__exportStar(require("@klyra/proto/src/codegen/cosmos/gov/v1/gov"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/clob/order"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/clob/tx"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/delaymsg/tx"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/perpetuals/tx"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/prices/tx"), exports);
__exportStar(require("@klyra/proto/src/codegen/google/protobuf/any"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/subaccounts/subaccount"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/sending/tx"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/sending/transfer"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/assets/genesis"), exports);
__exportStar(require("@klyra/proto/src/codegen/klyraprotocol/assets/asset"), exports);
