// @klyra/chain/dist/src/lib/subaccount.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SubaccountInfo = exports.WalletSubaccountInfo = void 0;
const shared_1 = require("@klyra/shared");
/**
 * @description A class that contains a wallet and a subaccount number
 * It's used to execute transactions on behalf of a subaccount
 */
class WalletSubaccountInfo {
    constructor(wallet, subaccountNumber = 0) {
        if (subaccountNumber < 0 || subaccountNumber > shared_1.MAX_SUBACCOUNT_NUMBER) {
            throw new Error(`Subaccount number must be between 0 and ${shared_1.MAX_SUBACCOUNT_NUMBER}`);
        }
        this.wallet = wallet;
        this.subaccountNumber = subaccountNumber;
    }
    get address() {
        const address = this.wallet.getAddress();
        if (address.length > 0) {
            return address;
        }
        throw new Error("Address not set");
    }
}
exports.WalletSubaccountInfo = WalletSubaccountInfo;
/**
 * @description A class that contains a subaccount number and an address
 * It's used to identify a subaccount, even when the wallet is not available
 *
 * With this class you can get the address of the subaccount, but you cannot sign transactions
 */
class SubaccountInfo {
    constructor(address, subaccountNumber = 0) {
        if (subaccountNumber < 0 || subaccountNumber > shared_1.MAX_SUBACCOUNT_NUMBER) {
            throw new Error(`Subaccount number must be between 0 and ${shared_1.MAX_SUBACCOUNT_NUMBER}`);
        }
        this.subaccountNumber = subaccountNumber;
        this.address = address;
    }
}
exports.SubaccountInfo = SubaccountInfo;
