// @klyra/proto/src/codegen/klyraprotocol/prices/market_price.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MarketPriceUpdates = exports.MarketPnlPriceUpdate = exports.MarketSpotPriceUpdate = exports.MarketPriceUpdate = exports.MarketSpotPrice = exports.MarketPrice = void 0;
const helpers_1 = require("../../helpers");
const _m0 = __importStar(require("protobufjs/minimal"));
function createBaseMarketPrice() {
    return {
        id: 0,
        exponent: 0,
        spotPrice: helpers_1.Long.UZERO,
        pnlPrice: helpers_1.Long.UZERO
    };
}
exports.MarketPrice = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== 0) {
            writer.uint32(8).uint32(message.id);
        }
        if (message.exponent !== 0) {
            writer.uint32(16).sint32(message.exponent);
        }
        if (!message.spotPrice.isZero()) {
            writer.uint32(24).uint64(message.spotPrice);
        }
        if (!message.pnlPrice.isZero()) {
            writer.uint32(32).uint64(message.pnlPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketPrice();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.uint32();
                    break;
                case 2:
                    message.exponent = reader.sint32();
                    break;
                case 3:
                    message.spotPrice = reader.uint64();
                    break;
                case 4:
                    message.pnlPrice = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMarketPrice();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : 0;
        message.exponent = (_b = object.exponent) !== null && _b !== void 0 ? _b : 0;
        message.spotPrice = object.spotPrice !== undefined && object.spotPrice !== null ? helpers_1.Long.fromValue(object.spotPrice) : helpers_1.Long.UZERO;
        message.pnlPrice = object.pnlPrice !== undefined && object.pnlPrice !== null ? helpers_1.Long.fromValue(object.pnlPrice) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMarketSpotPrice() {
    return {
        id: 0,
        exponent: 0,
        spotPrice: helpers_1.Long.UZERO
    };
}
exports.MarketSpotPrice = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.id !== 0) {
            writer.uint32(8).uint32(message.id);
        }
        if (message.exponent !== 0) {
            writer.uint32(16).sint32(message.exponent);
        }
        if (!message.spotPrice.isZero()) {
            writer.uint32(24).uint64(message.spotPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketSpotPrice();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.id = reader.uint32();
                    break;
                case 2:
                    message.exponent = reader.sint32();
                    break;
                case 3:
                    message.spotPrice = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a, _b;
        const message = createBaseMarketSpotPrice();
        message.id = (_a = object.id) !== null && _a !== void 0 ? _a : 0;
        message.exponent = (_b = object.exponent) !== null && _b !== void 0 ? _b : 0;
        message.spotPrice = object.spotPrice !== undefined && object.spotPrice !== null ? helpers_1.Long.fromValue(object.spotPrice) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMarketPriceUpdate() {
    return {
        marketId: 0,
        spotPrice: helpers_1.Long.UZERO,
        pnlPrice: helpers_1.Long.UZERO
    };
}
exports.MarketPriceUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.marketId !== 0) {
            writer.uint32(8).uint32(message.marketId);
        }
        if (!message.spotPrice.isZero()) {
            writer.uint32(16).uint64(message.spotPrice);
        }
        if (!message.pnlPrice.isZero()) {
            writer.uint32(24).uint64(message.pnlPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketPriceUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketId = reader.uint32();
                    break;
                case 2:
                    message.spotPrice = reader.uint64();
                    break;
                case 3:
                    message.pnlPrice = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMarketPriceUpdate();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : 0;
        message.spotPrice = object.spotPrice !== undefined && object.spotPrice !== null ? helpers_1.Long.fromValue(object.spotPrice) : helpers_1.Long.UZERO;
        message.pnlPrice = object.pnlPrice !== undefined && object.pnlPrice !== null ? helpers_1.Long.fromValue(object.pnlPrice) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMarketSpotPriceUpdate() {
    return {
        marketId: 0,
        spotPrice: helpers_1.Long.UZERO
    };
}
exports.MarketSpotPriceUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.marketId !== 0) {
            writer.uint32(8).uint32(message.marketId);
        }
        if (!message.spotPrice.isZero()) {
            writer.uint32(16).uint64(message.spotPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketSpotPriceUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketId = reader.uint32();
                    break;
                case 2:
                    message.spotPrice = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMarketSpotPriceUpdate();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : 0;
        message.spotPrice = object.spotPrice !== undefined && object.spotPrice !== null ? helpers_1.Long.fromValue(object.spotPrice) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMarketPnlPriceUpdate() {
    return {
        marketId: 0,
        pnlPrice: helpers_1.Long.UZERO
    };
}
exports.MarketPnlPriceUpdate = {
    encode(message, writer = _m0.Writer.create()) {
        if (message.marketId !== 0) {
            writer.uint32(8).uint32(message.marketId);
        }
        if (!message.pnlPrice.isZero()) {
            writer.uint32(16).uint64(message.pnlPrice);
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketPnlPriceUpdate();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketId = reader.uint32();
                    break;
                case 2:
                    message.pnlPrice = reader.uint64();
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMarketPnlPriceUpdate();
        message.marketId = (_a = object.marketId) !== null && _a !== void 0 ? _a : 0;
        message.pnlPrice = object.pnlPrice !== undefined && object.pnlPrice !== null ? helpers_1.Long.fromValue(object.pnlPrice) : helpers_1.Long.UZERO;
        return message;
    }
};
function createBaseMarketPriceUpdates() {
    return {
        marketPriceUpdates: []
    };
}
exports.MarketPriceUpdates = {
    encode(message, writer = _m0.Writer.create()) {
        for (const v of message.marketPriceUpdates) {
            exports.MarketPriceUpdate.encode(v, writer.uint32(10).fork()).ldelim();
        }
        return writer;
    },
    decode(input, length) {
        const reader = input instanceof _m0.Reader ? input : new _m0.Reader(input);
        let end = length === undefined ? reader.len : reader.pos + length;
        const message = createBaseMarketPriceUpdates();
        while (reader.pos < end) {
            const tag = reader.uint32();
            switch (tag >>> 3) {
                case 1:
                    message.marketPriceUpdates.push(exports.MarketPriceUpdate.decode(reader, reader.uint32()));
                    break;
                default:
                    reader.skipType(tag & 7);
                    break;
            }
        }
        return message;
    },
    fromPartial(object) {
        var _a;
        const message = createBaseMarketPriceUpdates();
        message.marketPriceUpdates = ((_a = object.marketPriceUpdates) === null || _a === void 0 ? void 0 : _a.map(e => exports.MarketPriceUpdate.fromPartial(e))) || [];
        return message;
    }
};
