// @klyra/chain/dist/src/auth/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getAccountFromUUID = getAccountFromUUID;
exports.getAccountFromCredentials = getAccountFromCredentials;
exports.getAccountFromSignedMessage = getAccountFromSignedMessage;
const bip32_1 = require("@scure/bip32");
const bip39_1 = require("@scure/bip39");
const keccak_1 = require("ethereum-cryptography/keccak");
const english_1 = require("@scure/bip39/wordlists/english");
const helpers_1 = require("../utils/helpers");
const local_wallet_1 = require("../lib/local-wallet");
const constants_1 = require("../constants");
async function getAccountFromUUID(uuid) {
    let buffer = new TextEncoder().encode(uuid);
    if (buffer.length < 65) {
        const padding = new Uint8Array(65 - buffer.length);
        buffer = new Uint8Array([...buffer, ...padding]);
    }
    if (buffer.length > 65) {
        buffer = buffer.slice(0, 65);
    }
    const adjustedUUID = Array.from(buffer)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join("");
    const { mnemonic, privateKey, publicKey } = getHDKeyFromMessage(adjustedUUID);
    // Create wallet from mnemonic
    const wallet = await local_wallet_1.LocalWallet.fromMnemonic(mnemonic, constants_1.BECH32_PREFIX);
    return {
        wallet,
        mnemonic,
        privateKey,
        publicKey,
    };
}
async function getAccountFromCredentials(username, pwd) {
    const usernameHash = (0, keccak_1.keccak256)(new TextEncoder().encode(username));
    const pwdHash = (0, keccak_1.keccak256)(new TextEncoder().encode(pwd));
    const concatenatedHashes = new Uint8Array([
        ...usernameHash,
        ...pwdHash,
        0x01,
    ]);
    const { mnemonic, privateKey, publicKey } = getHDKeyFromMessage(Array.from(concatenatedHashes)
        .map((b) => b.toString(16).padStart(2, "0"))
        .join(""));
    const wallet = await local_wallet_1.LocalWallet.fromMnemonic(mnemonic, constants_1.BECH32_PREFIX);
    return {
        wallet,
        mnemonic,
        privateKey,
        publicKey,
    };
}
async function getAccountFromSignedMessage(signedMessage) {
    const { mnemonic, privateKey, publicKey } = getHDKeyFromMessage(signedMessage);
    const wallet = await local_wallet_1.LocalWallet.fromMnemonic(mnemonic, constants_1.BECH32_PREFIX);
    return {
        wallet,
        mnemonic,
        privateKey,
        publicKey,
    };
}
function getHDKeyFromMessage(message) {
    var _a, _b;
    const buffer = new Uint8Array((_b = (_a = (0, helpers_1.stripHexPrefix)(message)
        .match(/.{1,2}/g)) === null || _a === void 0 ? void 0 : _a.map((byte) => parseInt(byte, 16))) !== null && _b !== void 0 ? _b : []);
    if (buffer.length !== 65) {
        throw new Error("Message must be 65 bytes");
    }
    // Remove the 'v' value by taking only the first 64 bytes of the message
    const rsValues = buffer.slice(0, 64);
    // Hash the 'r' and 's' values down to 32 bytes (256 bits) using Keccak-256
    const entropy = (0, keccak_1.keccak256)(rsValues);
    return exportMnemonicAndPrivateKey(entropy);
}
function exportMnemonicAndPrivateKey(entropy, path = "m/44'/118'/0'/0/0") {
    const mnemonic = (0, bip39_1.entropyToMnemonic)(entropy, english_1.wordlist);
    const { privateKey, publicKey } = deriveHDKeyFromMnemonic(mnemonic, path);
    return {
        mnemonic,
        privateKey,
        publicKey,
    };
}
function deriveHDKeyFromMnemonic(mnemonic, path) {
    const seed = (0, bip39_1.mnemonicToSeedSync)(mnemonic);
    const hdkey = bip32_1.HDKey.fromMasterSeed(seed);
    const derivedHdkey = hdkey.derive(path);
    if (!hdkey.privateKey) {
        throw new Error("null hd key");
    }
    return {
        privateKey: derivedHdkey.privateKey,
        publicKey: derivedHdkey.publicKey,
    };
}
