// @klyra/chain/dist/src/constants/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BECH32_PREFIX = exports.Messages = exports.MAX_UINT_32 = exports.GAS_MULTIPLIER = exports.TDAI_DECIMALS = exports.TDAI_DENOM = exports.PAGE_REQUEST = exports.DEFAULT_SEQUENCE = exports.DEFAULT_BROADCAST_OPTIONS = exports.SHORT_BLOCK_FORWARD = exports.SHORT_BLOCK_WINDOW = exports.DEFAULT_API_TIMEOUT = void 0;
const long_1 = __importDefault(require("long"));
// API options
exports.DEFAULT_API_TIMEOUT = 5000;
exports.SHORT_BLOCK_WINDOW = 20;
exports.SHORT_BLOCK_FORWARD = 3;
// Broadcast options
exports.DEFAULT_BROADCAST_OPTIONS = {
    pollIntervalMs: 300,
    timeoutMs: 8000,
};
// Validation options
exports.DEFAULT_SEQUENCE = 0;
// Querying
exports.PAGE_REQUEST = {
    key: new Uint8Array(),
    offset: long_1.default.UZERO,
    limit: long_1.default.MAX_UNSIGNED_VALUE,
    countTotal: true,
    reverse: false,
};
// Gas configs
exports.TDAI_DENOM = "utdai";
exports.TDAI_DECIMALS = 6;
// TODO: check what this is
exports.GAS_MULTIPLIER = 1.4;
exports.MAX_UINT_32 = 4294967295;
var messages_1 = require("./messages");
Object.defineProperty(exports, "Messages", { enumerable: true, get: function () { return __importDefault(messages_1).default; } });
exports.BECH32_PREFIX = "klyra";
