// @klyra/cortex/dist/src/__tests__/mocks/candles.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BATCH_UPDATE_MESSAGE = exports.SUBSCRIBED_MESSAGE = void 0;
const shared_1 = require("@klyra/shared");
exports.SUBSCRIBED_MESSAGE = {
    type: shared_1.MessageType.SUBSCRIBED,
    connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
    message_id: 5,
    channel: shared_1.MessageChannel.CANDLES,
    id: "ETH-USD/1MIN",
    contents: {
        candles: [
            {
                startedAt: "2024-09-25T14:47:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2632.6",
                high: "2632.6",
                open: "2632.6",
                close: "2632.6",
                baseTokenVolume: "0",
                usdVolume: "0",
                trades: 0,
                startingOpenInterest: "963.465",
                orderbookMidPriceOpen: "2623.05",
                orderbookMidPriceClose: "2623.05",
            },
            {
                startedAt: "2024-09-25T14:46:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2632.6",
                high: "2632.6",
                open: "2632.6",
                close: "2632.6",
                baseTokenVolume: "0",
                usdVolume: "0",
                trades: 0,
                startingOpenInterest: "963.465",
                orderbookMidPriceOpen: "2629.2",
                orderbookMidPriceClose: "2623.05",
            },
            {
                startedAt: "2024-09-25T14:45:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2632.6",
                high: "2632.6",
                open: "2632.6",
                close: "2632.6",
                baseTokenVolume: "0.004",
                usdVolume: "10.5304",
                trades: 2,
                startingOpenInterest: "963.465",
                orderbookMidPriceOpen: "2631.1",
                orderbookMidPriceClose: "2629.2",
            },
            {
                startedAt: "2024-09-25T14:44:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2632.6",
                high: "2632.6",
                open: "2632.6",
                close: "2632.6",
                baseTokenVolume: "0.004",
                usdVolume: "10.5304",
                trades: 2,
                startingOpenInterest: "963.463",
                orderbookMidPriceOpen: "2624.45",
                orderbookMidPriceClose: "2631.1",
            },
            {
                startedAt: "2024-09-25T14:43:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2634.6",
                high: "2634.6",
                open: "2634.6",
                close: "2634.6",
                baseTokenVolume: "0.008",
                usdVolume: "21.0768",
                trades: 4,
                startingOpenInterest: "963.459",
                orderbookMidPriceOpen: "2632.9",
                orderbookMidPriceClose: "2624.45",
            },
            {
                startedAt: "2024-09-25T14:42:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2634.5",
                high: "2634.5",
                open: "2634.5",
                close: "2634.5",
                baseTokenVolume: "0.008",
                usdVolume: "21.076",
                trades: 4,
                startingOpenInterest: "963.459",
                orderbookMidPriceOpen: "2624.5",
                orderbookMidPriceClose: "2632.9",
            },
            {
                startedAt: "2024-09-25T14:41:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2634",
                high: "2634.1",
                open: "2634",
                close: "2634.1",
                baseTokenVolume: "0.008",
                usdVolume: "21.0724",
                trades: 4,
                startingOpenInterest: "963.451",
                orderbookMidPriceOpen: "2631.6",
                orderbookMidPriceClose: "2624.5",
            },
            {
                startedAt: "2024-09-25T14:40:00.000Z",
                ticker: "ETH-USD",
                resolution: "1MIN",
                low: "2634",
                high: "2634",
                open: "2634",
                close: "2634",
                baseTokenVolume: "0.006",
                usdVolume: "15.804",
                trades: 3,
                startingOpenInterest: "963.451",
                orderbookMidPriceOpen: "2625.9",
                orderbookMidPriceClose: "2631.6",
            },
        ],
    },
};
exports.BATCH_UPDATE_MESSAGE = {
    type: shared_1.MessageType.CHANNEL_BATCH_DATA,
    connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
    message_id: 10,
    id: "ETH-USD/1MIN",
    channel: shared_1.MessageChannel.CANDLES,
    version: "1.0.0",
    contents: [
        {
            startedAt: "2024-09-25T14:47:00.000Z",
            ticker: "ETH-USD",
            resolution: "1MIN",
            low: "263.6",
            high: "263.6",
            open: "263.6",
            close: "263.6",
            baseTokenVolume: "0",
            usdVolume: "0",
            trades: 0,
            startingOpenInterest: "963.465",
            orderbookMidPriceOpen: "2623.05",
            orderbookMidPriceClose: "2629.25",
        },
        {
            startedAt: "2024-09-25T14:48:00.000Z",
            ticker: "ETH-USD",
            resolution: "1MIN",
            low: "2632.6",
            high: "2632.6",
            open: "2632.6",
            close: "2632.6",
            baseTokenVolume: "0",
            usdVolume: "0",
            trades: 0,
            startingOpenInterest: "963.465",
            orderbookMidPriceClose: "2629.25",
            orderbookMidPriceOpen: "2629.25",
        },
    ],
};
