// @klyra/cortex/dist/src/__tests__/mocks/markets.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TRADING_MARKET_UPDATE_MESSAGE = exports.ORACLE_PRICE_UPDATE_MESSAGE = exports.SUBSCRIBED_MESSAGE = void 0;
const shared_1 = require("@klyra/shared");
exports.SUBSCRIBED_MESSAGE = {
    type: shared_1.MessageType.SUBSCRIBED,
    connection_id: "c4773542-f710-4069-8211-83840c6b010d",
    message_id: 1,
    channel: shared_1.MessageChannel.MARKETS,
    contents: {
        markets: {
            "BTC-USD": {
                clobPairId: "0",
                ticker: "BTC-USD",
                status: "ACTIVE",
                spotPrice: "63360.41819",
                priceChange24H: "-395.10181",
                volume24H: "10860.1122",
                trades24H: 1663,
                nextFundingRate: "0",
                initialMarginFraction: "0.05",
                maintenanceMarginFraction: "0.03",
                openInterest: "50.3349",
                atomicResolution: -10,
                quantumConversionExponent: -9,
                tickSize: "1",
                stepSize: "0.0001",
                stepBaseQuantums: 1000000,
                subticksPerTick: 100000,
                marketType: "CROSS",
                openInterestLowerCap: "0",
                openInterestUpperCap: "0",
                baseOpenInterest: "50.3349",
            },
            "ETH-USD": {
                clobPairId: "1",
                ticker: "ETH-USD",
                status: "ACTIVE",
                spotPrice: "2582.364261",
                priceChange24H: "-55.325739",
                volume24H: "21223.3024",
                trades24H: 2109,
                nextFundingRate: "0",
                initialMarginFraction: "0.05",
                maintenanceMarginFraction: "0.03",
                openInterest: "964.107",
                atomicResolution: -9,
                quantumConversionExponent: -9,
                tickSize: "0.1",
                stepSize: "0.001",
                stepBaseQuantums: 1000000,
                subticksPerTick: 100000,
                marketType: "CROSS",
                openInterestLowerCap: "0",
                openInterestUpperCap: "0",
                baseOpenInterest: "964.107",
            },
        },
    },
};
exports.ORACLE_PRICE_UPDATE_MESSAGE = {
    type: shared_1.MessageType.CHANNEL_BATCH_DATA,
    connection_id: "c4773542-f710-4069-8211-83840c6b010d",
    message_id: 214,
    channel: shared_1.MessageChannel.MARKETS,
    version: "1.0.0",
    contents: [
        {
            oraclePrices: {
                "BTC-USD": { id: "0", spotPrice: "63360.41819" },
                "ETH-USD": { id: "1", spotPrice: "2582.364261" },
            },
        },
    ],
};
exports.TRADING_MARKET_UPDATE_MESSAGE = {
    type: shared_1.MessageType.CHANNEL_BATCH_DATA,
    connection_id: "c4773542-f710-4069-8211-83840c6b010d",
    message_id: 214,
    channel: shared_1.MessageChannel.MARKETS,
    version: "1.0.0",
    contents: [
        {
            trading: {
                "BTC-USD": {
                    id: "0",
                    trades24H: 10000,
                },
            },
        },
    ],
};
