// @klyra/cortex/dist/src/__tests__/mocks/orderbook.mock.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BATCH_UPDATE_MESSAGE = exports.SUBSCRIBED_MESSAGE = void 0;
const shared_1 = require("@klyra/shared");
exports.SUBSCRIBED_MESSAGE = {
    type: shared_1.MessageType.SUBSCRIBED,
    connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
    message_id: 4,
    channel: shared_1.MessageChannel.ORDERBOOK,
    id: "ETH-USD",
    contents: {
        bids: [
            { price: "2613.6", size: "0.415" },
            { price: "2612.1", size: "0.415" },
            { price: "2606.6", size: "7.673" },
            { price: "2606.5", size: "15.346" },
            { price: "2606.4", size: "46.038" },
            { price: "2606", size: "2.301" },
            { price: "2605.9", size: "0.768" },
            { price: "2605.8", size: "99.775" },
            { price: "2605.7", size: "0.768" },
            { price: "2605.6", size: "8.444" },
        ],
        asks: [
            { price: "2629.9", size: "0.415" },
            { price: "2633.6", size: "0.415" },
            { price: "2637.9", size: "0.758" },
            { price: "2638.7", size: "0.758" },
            { price: "2638.8", size: "0.758" },
            { price: "2638.9", size: "0.758" },
            { price: "2639.1", size: "1.516" },
            { price: "2639.5", size: "3.032" },
            { price: "2639.7", size: "2.274" },
            { price: "2639.8", size: "0.758" },
        ],
    },
};
exports.BATCH_UPDATE_MESSAGE = {
    type: shared_1.MessageType.CHANNEL_BATCH_DATA,
    connection_id: "54480a41-95fe-4253-9b39-44a465690af2",
    message_id: 7,
    id: "ETH-USD",
    channel: shared_1.MessageChannel.ORDERBOOK,
    version: "1.0.0",
    contents: [
        { bids: [["2613.6", "140"]] },
        { asks: [["2629.9", "140"]] },
        { asks: [["32", "100"]], bids: [["26032", "100"]] },
    ],
};
